From 00b0f489ea209c0d66acc5e3154385a5c7d5b8c5 Mon Sep 17 00:00:00 2001
From: Vaciller <wyc_19533480830@outlook.com>
Date: Fri, 10 Jul 2026 15:07:08 +0800
Subject: [PATCH] feat: add Rockchip HDMI camera status sync via
 rockchip.hardware.hdmi@1.0

Change-Id: Ib6971c50e4908373863ee546a841b0052bfd6f8f

diff --git a/services/camera/libcameraservice/Android.bp b/services/camera/libcameraservice/Android.bp
index 981c56942c..e20bda101e 100644
--- a/services/camera/libcameraservice/Android.bp
+++ b/services/camera/libcameraservice/Android.bp
@@ -165,6 +165,7 @@ cc_library_shared {
         "android.hardware.camera.device@3.7",
         "android.hardware.camera.device-V1-ndk",
         "media_permission-aidl-cpp",
+        "rockchip.hardware.hdmi@1.0",
     ],
 
     static_libs: [
@@ -175,6 +176,14 @@ cc_library_shared {
         "libcameraservice_device_independent",
     ],
 
+    product_variables: {
+       cameraSupportHDMI: {
+           cppflags:[
+               "-DHDMI_ENABLE",
+           ],
+       }
+    },
+
     export_shared_lib_headers: [
         "libbinder",
         "libactivitymanager_aidl",
diff --git a/services/camera/libcameraservice/CameraService.cpp b/services/camera/libcameraservice/CameraService.cpp
old mode 100644
new mode 100755
index a6cd5cdb17..ac423a9a05
--- a/services/camera/libcameraservice/CameraService.cpp
+++ b/services/camera/libcameraservice/CameraService.cpp
@@ -77,7 +77,9 @@
 #include "utils/TagMonitor.h"
 #include "utils/CameraThreadState.h"
 #include "utils/CameraServiceProxyWrapper.h"
-
+#ifdef HDMI_ENABLE
+#include <rockchip/hardware/hdmi/1.0/IHdmi.h>
+#endif
 namespace {
     const char* kPermissionServiceName = "permission";
 }; // namespace anonymous
@@ -444,6 +446,20 @@ void CameraService::onDeviceStatusChanged(const String8& id,
             addStates(id);
 
             updateStatus(newStatus, id);
+#ifdef HDMI_ENABLE
+            sp<rockchip::hardware::hdmi::V1_0::IHdmi> client = rockchip::hardware::hdmi::V1_0::IHdmi::getService();
+            ::android::hardware::hidl_string deviceId;
+            if(client.get()!= nullptr){
+                client->getHdmiDeviceId( [&](const ::android::hardware::hidl_string &id){
+                        deviceId = id.c_str();
+                        ALOGD("cameraId:%s",id.c_str());
+                });
+            }
+            if(client.get()!= nullptr && strstr(id.string(),deviceId.c_str())){
+                client->onStatusChange((uint32_t)newHalStatus);
+                ALOGD("onStatusChange:%d",newHalStatus);
+            }
+#endif
         } else {
             ALOGE("%s: Bad camera ID %s", __FUNCTION__, id.string());
         }
@@ -490,6 +506,20 @@ void CameraService::onDeviceStatusChanged(const String8& id,
         }
         updateStatus(newStatus, id);
     }
+#ifdef HDMI_ENABLE
+    sp<rockchip::hardware::hdmi::V1_0::IHdmi> client = rockchip::hardware::hdmi::V1_0::IHdmi::getService();
+    ::android::hardware::hidl_string deviceId;
+    if(client.get()!= nullptr){
+        client->getHdmiDeviceId( [&](const ::android::hardware::hidl_string &id){
+                deviceId = id.c_str();
+                ALOGD("cameraId:%s",id.c_str());
+        });
+    }
+    if(client.get()!= nullptr && strstr(id.string(),deviceId.c_str())){
+        client->onStatusChange((uint32_t)newHalStatus);
+        ALOGD("onStatusChange:%d",newHalStatus);
+    }
+#endif
 }
 
 void CameraService::onDeviceStatusChanged(const String8& id,
@@ -1837,7 +1867,16 @@ Status CameraService::connectHelper(const sp<CALLBACK>& cameraCb, const String8&
     ALOGI("CameraService::connect call (PID %d \"%s\", camera ID %s) and "
             "Camera API version %d", packagePid, clientName8.string(), cameraId.string(),
             static_cast<int>(effectiveApiLevel));
-
+#ifdef HDMI_ENABLE
+    sp<rockchip::hardware::hdmi::V1_0::IHdmi> hdmi= rockchip::hardware::hdmi::V1_0::IHdmi::getService();
+    if(hdmi.get()!= nullptr){
+        ::android::hardware::hidl_string deviceId = cameraId.string();
+        rockchip::hardware::hdmi::V1_0::HdmiAudioStatus audioStatus;
+        audioStatus.status = 1;
+        audioStatus.deviceId = deviceId;
+        hdmi->onAudioChange(audioStatus);
+    }
+#endif
     nsecs_t openTimeNs = systemTime();
 
     sp<CLIENT> client = nullptr;
diff --git a/services/camera/libcameraservice/device3/Camera3Device.cpp b/services/camera/libcameraservice/device3/Camera3Device.cpp
old mode 100644
new mode 100755
index e55d724b9f..9e2c2a69c5
--- a/services/camera/libcameraservice/device3/Camera3Device.cpp
+++ b/services/camera/libcameraservice/device3/Camera3Device.cpp
@@ -65,6 +65,10 @@
 #include "utils/TraceHFR.h"
 #include "utils/CameraServiceProxyWrapper.h"
 
+#ifdef HDMI_ENABLE
+#include <rockchip/hardware/hdmi/1.0/IHdmi.h>
+#endif
+
 #include <algorithm>
 #include <tuple>
 
@@ -249,7 +253,7 @@ status_t Camera3Device::disconnect() {
 
 status_t Camera3Device::disconnectImpl() {
     ATRACE_CALL();
-    ALOGI("%s: E", __FUNCTION__);
+    ALOGI("%s: E deviceId:%s", __FUNCTION__,mId.c_str());
 
     status_t res = OK;
     std::vector<wp<Camera3StreamInterface>> streams;
@@ -325,7 +329,16 @@ status_t Camera3Device::disconnectImpl() {
             mStatusTracker.clear();
             interface = mInterface.get();
         }
-
+#ifdef HDMI_ENABLE
+    sp<rockchip::hardware::hdmi::V1_0::IHdmi> hdmi= rockchip::hardware::hdmi::V1_0::IHdmi::getService();
+    if(hdmi.get()!= nullptr){
+         ::android::hardware::hidl_string deviceId = mId.c_str();
+        rockchip::hardware::hdmi::V1_0::HdmiAudioStatus audioStatus;
+        audioStatus.status = 0;
+        audioStatus.deviceId = deviceId;
+        hdmi->onAudioChange(audioStatus);
+    }
+#endif
         // Call close without internal mutex held, as the HAL close may need to
         // wait on assorted callbacks,etc, to complete before it can return.
         mCameraServiceWatchdog->WATCH(interface->close());
diff --git a/services/camera/libcameraservice/device3/Camera3OutputStream.cpp b/services/camera/libcameraservice/device3/Camera3OutputStream.cpp
index ef12b64972..b086d90549 100644
--- a/services/camera/libcameraservice/device3/Camera3OutputStream.cpp
+++ b/services/camera/libcameraservice/device3/Camera3OutputStream.cpp
@@ -458,7 +458,7 @@ status_t Camera3OutputStream::returnBufferCheckedLocked(
         // Fix CameraBlob id type discrepancy between HIDL and AIDL, details : http://b/229688810
         if (getFormat() == HAL_PIXEL_FORMAT_BLOB && getDataSpace() == HAL_DATASPACE_V0_JFIF) {
             if (mIPCTransport == IPCTransport::HIDL) {
-                fixUpHidlJpegBlobHeader(anwBuffer, anwReleaseFence);
+                //fixUpHidlJpegBlobHeader(anwBuffer, anwReleaseFence);
             }
             // If this is a JPEG output, and image dump mask is set, save image to
             // disk.
