From 94c43fd2d76093d07dd9f7711bcba59e4b9d7681 Mon Sep 17 00:00:00 2001
From: Vaciller <wyc_19533480830@outlook.com>
Date: Fri, 10 Jul 2026 17:46:34 +0800
Subject: [PATCH] feat: add AFBC crop support to NdkImageReader

Change-Id: Ia0a576845a201ca0de8371e3c19582328f215fe6

diff --git a/media/ndk/NdkImageReader.cpp b/media/ndk/NdkImageReader.cpp
index 92704995e0..3d78bf7f6b 100644
--- a/media/ndk/NdkImageReader.cpp
+++ b/media/ndk/NdkImageReader.cpp
@@ -451,7 +451,8 @@ AImageReader::acquireImageLocked(/*out*/AImage** image, /*out*/int* acquireFence
         Point lt = buffer->mCrop.leftTop();
         if (lt.x != 0 || lt.y != 0) {
             ALOGE("Crop left top corner [%d, %d] not at origin", lt.x, lt.y);
-            return AMEDIA_ERROR_UNKNOWN;
+            // for support afbc mode
+            //return AMEDIA_ERROR_UNKNOWN;
         }
 
         // Check if the producer buffer configurations match what ImageReader configured.
