From 2889c44f8caac9e7d0bf58e034f255fe71e02393 Mon Sep 17 00:00:00 2001
From: Vaciller <wyc_19533480830@outlook.com>
Date: Fri, 10 Jul 2026 17:14:08 +0800
Subject: [PATCH] feat: add boot video playback as boot animation

Change-Id: I571ef665c565b3241d210ce801a8bc0c8b87e68c

diff --git a/cmds/bootanimation/Android.bp b/cmds/bootanimation/Android.bp
index 3534624a58a2..f1ed94f50d29 100644
--- a/cmds/bootanimation/Android.bp
+++ b/cmds/bootanimation/Android.bp
@@ -70,6 +70,7 @@ cc_library_shared {
     shared_libs: [
         "libui",
         "libjnigraphics",
+        "libmedia",
         "libEGL",
         "libGLESv2",
         "libgui",
diff --git a/cmds/bootanimation/BootAnimation.cpp b/cmds/bootanimation/BootAnimation.cpp
index 53a84bd1b21e..23175216b137 100644
--- a/cmds/bootanimation/BootAnimation.cpp
+++ b/cmds/bootanimation/BootAnimation.cpp
@@ -57,6 +57,9 @@
 #include <EGL/eglext.h>
 
 #include "BootAnimation.h"
+#include "audioplay.h"
+#include <media/mediaplayer.h>
+#include <media/IMediaHTTPService.h>
 
 #define ANIM_PATH_MAX 255
 #define STR(x)   #x
@@ -84,6 +87,9 @@ static constexpr const char* SYSTEM_USERSPACE_REBOOT_ANIMATION_FILE = "/system/m
 static const char BOOTANIM_DATA_DIR_PATH[] = "/data/bootanim";
 static const char BOOTANIM_TIME_DIR_NAME[] = "time";
 static const char BOOTANIM_TIME_DIR_PATH[] = "/data/bootanim/time";
+//support boot video
+static const char PRODUCT_BOOTVIDEO_FILE[] = "/product/media/bootanimation.ts";
+
 static const char CLOCK_FONT_ASSET[] = "images/clock_font.png";
 static const char CLOCK_FONT_ZIP_NAME[] = "clock_font.png";
 static const char PROGRESS_FONT_ASSET[] = "images/progress_font.png";
@@ -714,6 +720,23 @@ void BootAnimation::findBootAnimationFile() {
     bool encryptedAnimation = atoi(decrypt) != 0 ||
         !strcmp("trigger_restart_min_framework", decrypt);
 
+    //add for boot video
+    mVideoAnimation = false;
+    if (access(PRODUCT_BOOTVIDEO_FILE, R_OK) == 0) {
+        mVideoFile = (char*)PRODUCT_BOOTVIDEO_FILE;
+    }
+    property_get("persist.sys.bootvideo.enable", decrypt, "false");
+    char value[PROPERTY_VALUE_MAX];
+    property_get("persist.sys.bootvideo.showtime", value, "-1");
+    if (mVideoFile != NULL && !strcmp(decrypt, "true") && (atoi(value) != 0)) {
+        mVideoAnimation = true;
+        ALOGD("mVideoAnimation is true");
+    } else {
+        ALOGD("bootvideo:No boot video animation,EXIT_VIDEO_NAME:%s, bootvideo.showtime:%s\n",
+            decrypt, value);
+    }
+    //add end
+
     if (!mShuttingDown && encryptedAnimation) {
         static const std::vector<std::string> encryptedBootFiles = {
             PRODUCT_ENCRYPTED_BOOTANIMATION_FILE, SYSTEM_ENCRYPTED_BOOTANIMATION_FILE,
@@ -814,14 +837,21 @@ void BootAnimation::initShaders() {
 bool BootAnimation::threadLoop() {
     bool result;
     initShaders();
+    //add for boot video function
+    mStartbootanimaTime = 0;
+    mBootVideoTime = -1;
 
-    // We have no bootanimation file, so we use the stock android logo
-    // animation.
-    if (mZipFileName.isEmpty()) {
-        ALOGD("No animation file");
-        result = android();
+    if (mVideoAnimation) {
+        result = video();
     } else {
-        result = movie();
+        // We have no bootanimation file, so we use the stock android logo
+        // animation.
+        if (mZipFileName.isEmpty()) {
+            ALOGD("No animation file");
+            result = android();
+        } else {
+            result = movie();
+        }
     }
 
     mCallbacks->shutdown();
@@ -908,8 +938,21 @@ void BootAnimation::checkExit() {
     char value[PROPERTY_VALUE_MAX];
     property_get(EXIT_PROP_NAME, value, "0");
     int exitnow = atoi(value);
+    //add for boot video function
+    property_get("persist.sys.bootvideo.enable", value, "false");
+    const nsecs_t realBootanimaTime = systemTime() - mStartbootanimaTime;
     if (exitnow) {
-        requestExit();
+        //add for boot video function
+        if (!strcmp(value,"true")) {
+            if ((ns2ms(realBootanimaTime)/1000) > mBootVideoTime) {
+                ALOGD("checkExit,requestExit for bootvideo");
+                //close bootvolume for audioflinger
+                property_set("sys.bootvideo.closed", "1");
+                requestExit();
+            }
+        } else {
+            requestExit();
+        }
     }
 }
 
@@ -1924,6 +1967,83 @@ status_t BootAnimation::TimeCheckThread::readyToRun() {
     return NO_ERROR;
 }
 
+//add for boot video
+bool BootAnimation::video() {
+    const bool LOOP = false;
+    const float CHECK_DELAY = 500*1000;//500ms
+    int duration = 0;
+    char delay[64];
+    sp<IMediaHTTPService> httpService;
+
+    char value[PROPERTY_VALUE_MAX];
+    property_get("persist.sys.bootvideo.showtime", value, "-1");
+    int bootvideo_time = atoi(value);//s
+    if(bootvideo_time > 120)
+        bootvideo_time = 120;
+
+    //add delaytime set ,default 0.
+    property_get("persist.sys.bootvideo.delaytime", value, "0");
+    int delay_new = atoi(value);//s
+    usleep(CHECK_DELAY * 2 * delay_new);
+
+    sp<MediaPlayer> mp = new MediaPlayer();
+    // create the native surface
+    sp<SurfaceControl> control = session()->createSurface(String8("BootAnimation_video"),
+         mWidth, mHeight, PIXEL_FORMAT_RGB_565);
+    SurfaceComposerClient::Transaction t;
+    t.setLayer(control, 0x40000001)
+        .apply();
+    sp<Surface> surface = control->getSurface();
+    mp->setDataSource(httpService, mVideoFile, NULL);
+    mp->setLooping(LOOP);
+    mp->setVideoSurfaceTexture(surface->getIGraphicBufferProducer());
+    mp->prepare();
+
+    //set bootvideo volume
+    property_get("persist.sys.bootvideo.vol", value, "-1");
+    if (strcmp(value, "-1")) {
+        float vol = 1.00f * atoi(value) / 100;//max 100
+        ALOGD("bootVideo vol=%f", vol);
+        mp->setVolume(vol, vol);
+    }
+
+    mp->getDuration(&duration);//persist.sys.pic_time
+    if (bootvideo_time > 0) {
+        sprintf(delay, "%d", bootvideo_time);
+    } else if (bootvideo_time == -2) {
+        sprintf(delay, "%d", (duration / 1000) + 1);
+    } else if (bootvideo_time == -1) {
+        sprintf(delay, "%d", 0);
+    }
+    //property_set("persist.sys.pic_time", delay);
+    mBootVideoTime = atoi(delay);
+    ALOGD("bootvideo:vol=%s,showtime=%d, duration=%d, delay=%s\n",
+        value, bootvideo_time, duration, delay);
+
+    mp->start();
+    mStartbootanimaTime = systemTime();
+    while (true) {
+        const nsecs_t realVideoTime = systemTime() - mStartbootanimaTime;
+        checkExit();
+        property_set("sys.bootvideo.closed", "0");
+        usleep(CHECK_DELAY);
+        if (!mp->isPlaying() || (((ns2ms(realVideoTime) / 1000) > bootvideo_time)
+            && (bootvideo_time > -1))) {
+            mp->pause();
+        }
+        if (exitPending()) {
+            ALOGD("bootvideo: stop bootanimation");
+            break;
+        }
+    }
+    property_set("sys.bootvideo.closed", "1");
+    mp->stop();
+    mp->reset();
+    surface.clear();
+    control.clear();
+    mp = NULL;
+    return false;
+}
 // ---------------------------------------------------------------------------
 
 } // namespace android
diff --git a/cmds/bootanimation/BootAnimation.h b/cmds/bootanimation/BootAnimation.h
index 8683b71a3762..64210424081c 100644
--- a/cmds/bootanimation/BootAnimation.h
+++ b/cmds/bootanimation/BootAnimation.h
@@ -183,6 +183,8 @@ private:
     void initShaders();
     bool android();
     bool movie();
+    //add for bootvideo
+    bool video();
     void drawText(const char* str, const Font& font, bool bold, int* x, int* y);
     void drawClock(const Font& font, const int xPos, const int yPos);
     void drawProgress(int percent, const Font& font, const int xPos, const int yPos);
@@ -236,6 +238,10 @@ private:
     bool        mShuttingDown;
     bool        mDynamicColorsApplied = false;
     String8     mZipFileName;
+    bool mVideoAnimation;
+    char* mVideoFile;
+    nsecs_t mStartbootanimaTime;
+    int  mBootVideoTime;
     SortedVector<String8> mLoadedFiles;
     sp<TimeCheckThread> mTimeCheckThread = nullptr;
     sp<Callbacks> mCallbacks;
