From fa2ff3f9d1440e0418cbf946f477d1a8179630d0 Mon Sep 17 00:00:00 2001
From: Vaciller <wyc_19533480830@outlook.com>
Date: Fri, 10 Jul 2026 17:14:44 +0800
Subject: [PATCH] feat: add HDMI-IN capture and stream buffer producer in JNI

Change-Id: I6ab32602115576c6c8338dba2b1a0eb39a1d6dff

diff --git a/services/core/jni/com_android_server_tv_TvInputHal.cpp b/services/core/jni/com_android_server_tv_TvInputHal.cpp
index a5311f33eb36..f42601afcfb1 100644
--- a/services/core/jni/com_android_server_tv_TvInputHal.cpp
+++ b/services/core/jni/com_android_server_tv_TvInputHal.cpp
@@ -16,7 +16,7 @@
 
 #define LOG_TAG "TvInputHal"
 
-//#define LOG_NDEBUG 0
+#define LOG_NDEBUG 0
 
 #include "android_os_MessageQueue.h"
 #include "android_runtime/AndroidRuntime.h"
@@ -24,9 +24,9 @@
 #include <nativehelper/JNIHelp.h>
 #include "jni.h"
 
-#include <android/hardware/tv/input/1.0/ITvInputCallback.h>
-#include <android/hardware/tv/input/1.0/ITvInput.h>
-#include <android/hardware/tv/input/1.0/types.h>
+#include <rockchip/hardware/tv/input/1.0/ITvInputCallback.h>
+#include <rockchip/hardware/tv/input/1.0/ITvInput.h>
+#include <rockchip/hardware/tv/input/1.0/types.h>
 #include <gui/Surface.h>
 #include <utils/Errors.h>
 #include <utils/KeyedVector.h>
@@ -34,21 +34,31 @@
 #include <utils/Looper.h>
 #include <utils/NativeHandle.h>
 #include <hardware/tv_input.h>
-
+#include <hardware/gralloc_rockchip.h>
+#include <cutils/properties.h>
+#define LOGV LOGD
 using ::android::hardware::audio::common::V2_0::AudioDevice;
-using ::android::hardware::tv::input::V1_0::ITvInput;
-using ::android::hardware::tv::input::V1_0::ITvInputCallback;
 using ::android::hardware::tv::input::V1_0::Result;
-using ::android::hardware::tv::input::V1_0::TvInputDeviceInfo;
 using ::android::hardware::tv::input::V1_0::TvInputEvent;
-using ::android::hardware::tv::input::V1_0::TvInputEventType;
 using ::android::hardware::tv::input::V1_0::TvInputType;
-using ::android::hardware::tv::input::V1_0::TvStreamConfig;
+using ::rockchip::hardware::tv::input::V1_0::ITvInput;
+using ::rockchip::hardware::tv::input::V1_0::ITvInputCallback;
+using ::rockchip::hardware::tv::input::V1_0::TvInputDeviceInfo;
+using ::rockchip::hardware::tv::input::V1_0::TvInputEventExt;
+using ::rockchip::hardware::tv::input::V1_0::TvInputEventType;
+using ::rockchip::hardware::tv::input::V1_0::TvStreamConfig;
+using ::rockchip::hardware::tv::input::V1_0::PreviewBuffer;
+using ::rockchip::hardware::tv::input::V1_0::PreviewRequest;
+using ::rockchip::hardware::tv::input::V1_0::PrivAppCmdInfo;
+using ::rockchip::hardware::tv::input::V1_0::PrivAppCmdBundle;
 using ::android::hardware::Return;
 using ::android::hardware::Void;
 using ::android::hardware::hidl_vec;
 using ::android::hardware::hidl_string;
 
+using ITvInputDeviceInfo = ::rockchip::hardware::tv::input::V1_0::TvInputDeviceInfo;
+using ITvInputEventType = ::android::hardware::tv::input::V1_0::TvInputEventType;
+
 namespace android {
 
 static struct {
@@ -56,6 +66,7 @@ static struct {
     jmethodID deviceUnavailable;
     jmethodID streamConfigsChanged;
     jmethodID firstFrameCaptured;
+    jmethodID privCmdToApp;
 } gTvInputHalClassInfo;
 
 static struct {
@@ -87,165 +98,23 @@ static struct {
     jmethodID build;
 } gTvInputHardwareInfoBuilderClassInfo;
 
-////////////////////////////////////////////////////////////////////////////////
-
-class BufferProducerThread : public Thread {
-public:
-    BufferProducerThread(tv_input_device_t* device, int deviceId, const tv_stream_t* stream);
-
-    virtual status_t readyToRun();
-
-    void setSurface(const sp<Surface>& surface);
-    void onCaptured(uint32_t seq, bool succeeded);
-    void shutdown();
-
-private:
-    Mutex mLock;
-    Condition mCondition;
-    sp<Surface> mSurface;
-    tv_input_device_t* mDevice;
-    int mDeviceId;
-    tv_stream_t mStream;
-    sp<ANativeWindowBuffer_t> mBuffer;
-    enum {
-        CAPTURING,
-        CAPTURED,
-        RELEASED,
-    } mBufferState;
-    uint32_t mSeq;
-    bool mShutdown;
-
-    virtual bool threadLoop();
-
-    void setSurfaceLocked(const sp<Surface>& surface);
+enum {
+    BUFF_STATUS_QUEUED,
+    BUFF_STATUS_DEQUEUED,
+    BUFF_STATUS_QUEUED_FAILED,
+    BUFF_STATUS_DEQUEUED_FAILED,
 };
 
-BufferProducerThread::BufferProducerThread(
-        tv_input_device_t* device, int deviceId, const tv_stream_t* stream)
-    : Thread(false),
-      mDevice(device),
-      mDeviceId(deviceId),
-      mBuffer(NULL),
-      mBufferState(RELEASED),
-      mSeq(0u),
-      mShutdown(false) {
-    memcpy(&mStream, stream, sizeof(mStream));
-}
-
-status_t BufferProducerThread::readyToRun() {
-    sp<ANativeWindow> anw(mSurface);
-    status_t err = native_window_set_usage(anw.get(), mStream.buffer_producer.usage);
-    if (err != NO_ERROR) {
-        return err;
-    }
-    err = native_window_set_buffers_dimensions(
-            anw.get(), mStream.buffer_producer.width, mStream.buffer_producer.height);
-    if (err != NO_ERROR) {
-        return err;
-    }
-    err = native_window_set_buffers_format(anw.get(), mStream.buffer_producer.format);
-    if (err != NO_ERROR) {
-        return err;
-    }
-    return NO_ERROR;
-}
-
-void BufferProducerThread::setSurface(const sp<Surface>& surface) {
-    Mutex::Autolock autoLock(&mLock);
-    setSurfaceLocked(surface);
-}
-
-void BufferProducerThread::setSurfaceLocked(const sp<Surface>& surface) {
-    if (surface == mSurface) {
-        return;
-    }
-
-    if (mBufferState == CAPTURING) {
-        mDevice->cancel_capture(mDevice, mDeviceId, mStream.stream_id, mSeq);
-    }
-    while (mBufferState == CAPTURING) {
-        status_t err = mCondition.waitRelative(mLock, s2ns(1));
-        if (err != NO_ERROR) {
-            ALOGE("error %d while wating for buffer state to change.", err);
-            break;
-        }
-    }
-    mBuffer.clear();
-    mBufferState = RELEASED;
-
-    mSurface = surface;
-    mCondition.broadcast();
-}
-
-void BufferProducerThread::onCaptured(uint32_t seq, bool succeeded) {
-    Mutex::Autolock autoLock(&mLock);
-    if (seq != mSeq) {
-        ALOGW("Incorrect sequence value: expected %u actual %u", mSeq, seq);
-    }
-    if (mBufferState != CAPTURING) {
-        ALOGW("mBufferState != CAPTURING : instead %d", mBufferState);
-    }
-    if (succeeded) {
-        mBufferState = CAPTURED;
-    } else {
-        mBuffer.clear();
-        mBufferState = RELEASED;
-    }
-    mCondition.broadcast();
-}
+typedef struct tvhal_preview_buff {
+    uint64_t buffId;
+    int bufferFenceFd;
+    int buffStatus;
+    sp<ANativeWindowBuffer_t> anwbPtr;
+    sp<GraphicBuffer> mGraphicBuffer;
+} tvhal_preview_buff_t;
 
-void BufferProducerThread::shutdown() {
-    Mutex::Autolock autoLock(&mLock);
-    mShutdown = true;
-    setSurfaceLocked(NULL);
-    requestExitAndWait();
-}
-
-bool BufferProducerThread::threadLoop() {
-    Mutex::Autolock autoLock(&mLock);
-
-    status_t err = NO_ERROR;
-    if (mSurface == NULL) {
-        err = mCondition.waitRelative(mLock, s2ns(1));
-        // It's OK to time out here.
-        if (err != NO_ERROR && err != TIMED_OUT) {
-            ALOGE("error %d while wating for non-null surface to be set", err);
-            return false;
-        }
-        return true;
-    }
-    sp<ANativeWindow> anw(mSurface);
-    while (mBufferState == CAPTURING) {
-        err = mCondition.waitRelative(mLock, s2ns(1));
-        if (err != NO_ERROR) {
-            ALOGE("error %d while wating for buffer state to change.", err);
-            return false;
-        }
-    }
-    if (mBufferState == CAPTURED && anw != NULL) {
-        err = anw->queueBuffer(anw.get(), mBuffer.get(), -1);
-        if (err != NO_ERROR) {
-            ALOGE("error %d while queueing buffer to surface", err);
-            return false;
-        }
-        mBuffer.clear();
-        mBufferState = RELEASED;
-    }
-    if (mBuffer == NULL && !mShutdown && anw != NULL) {
-        ANativeWindowBuffer_t* buffer = NULL;
-        err = native_window_dequeue_buffer_and_wait(anw.get(), &buffer);
-        if (err != NO_ERROR) {
-            ALOGE("error %d while dequeueing buffer to surface", err);
-            return false;
-        }
-        mBuffer = buffer;
-        mBufferState = CAPTURING;
-        mDevice->request_capture(mDevice, mDeviceId, mStream.stream_id,
-                                 buffer->handle, ++mSeq);
-    }
+////////////////////////////////////////////////////////////////////////////////
 
-    return true;
-}
 
 ////////////////////////////////////////////////////////////////////////////////
 
@@ -255,6 +124,7 @@ public:
 
     static JTvInputHal* createInstance(JNIEnv* env, jobject thiz, const sp<Looper>& looper);
 
+    int privCmdFromApp(const PrivAppCmdInfo& cmdInfo);
     int addOrUpdateStream(int deviceId, int streamId, const sp<Surface>& surface);
     int removeStream(int deviceId, int streamId);
     const hidl_vec<TvStreamConfig> getStreamConfigs(int deviceId);
@@ -262,9 +132,46 @@ public:
     void onDeviceAvailable(const TvInputDeviceInfo& info);
     void onDeviceUnavailable(int deviceId);
     void onStreamConfigurationsChanged(int deviceId, int cableConnectionStatus);
-    void onCaptured(int deviceId, int streamId, uint32_t seq, bool succeeded);
+    void onCaptured(int deviceId, int streamId, uint64_t buffId, int buffSeq, buffer_handle_t handle, bool succeeded);
+    void onPrivCmdToApp(int deviceId, const PrivAppCmdInfo& cmdInfo);
 
+    std::vector<PreviewBuffer> mPreviewBuffer;
 private:
+    class BufferProducerThread : public Thread {
+    public:
+        BufferProducerThread(JTvInputHal* hal, sp<ITvInput> tvinput, int deviceId);
+
+        int initPreviewBuffPoll(const sp<Surface>& surface, const tv_stream_config_ext_t* stream);
+        void setSurface(const sp<Surface>& surface);
+        void onCaptured(uint64_t buffId, int buffSeq, buffer_handle_t handle, bool succeeded);
+        void shutdown();
+
+    private:
+        Mutex mLock;
+        Condition mCondition;
+        sp<Surface> mSurface;
+        JTvInputHal* mParentHal;
+        sp<ITvInput> mTvInputPtr;
+        int mDeviceId;
+        tv_stream_config_ext mStream;
+        std::vector<tvhal_preview_buff_t> mTvHalPreviewBuff;
+        sp<ANativeWindowBuffer_t> mBuffer;
+        enum {
+            CAPTURING,
+            CAPTURED,
+            RELEASED,
+        } mBufferState;
+        uint32_t mSeq;
+        uint64_t mCurrBuffId;
+        bool mFirstCaptured;
+        bool mShutdown;
+        bool mAlreadyShowSignal;
+
+        virtual status_t readyToRun();
+        virtual bool threadLoop();
+        status_t configPreviewBuff();
+        void surfaceBuffDestory();
+    };
     // Connection between a surface and a stream.
     class Connection {
     public:
@@ -275,18 +182,19 @@ private:
 
         // Only valid when mStreamType == TV_STREAM_TYPE_INDEPENDENT_VIDEO_SOURCE
         sp<NativeHandle> mSourceHandle;
+        sp<NativeHandle> mCancelSourceHandle = NULL;
         // Only valid when mStreamType == TV_STREAM_TYPE_BUFFER_PRODUCER
         sp<BufferProducerThread> mThread;
     };
 
     class NotifyHandler : public MessageHandler {
     public:
-        NotifyHandler(JTvInputHal* hal, const TvInputEvent& event);
+        NotifyHandler(JTvInputHal* hal, const TvInputEventExt& extEvent);
 
         virtual void handleMessage(const Message& message);
 
     private:
-        TvInputEvent mEvent;
+        TvInputEventExt mEvent;
         JTvInputHal* mHal;
     };
 
@@ -294,6 +202,7 @@ private:
     public:
         explicit TvInputCallback(JTvInputHal* hal);
         Return<void> notify(const TvInputEvent& event) override;
+        Return<void> notify_ext(const TvInputEventExt& extEvent) override;
     private:
         JTvInputHal* mHal;
     };
@@ -317,11 +226,12 @@ JTvInputHal::JTvInputHal(JNIEnv* env, jobject thiz, sp<ITvInput> tvInput,
     mTvInput = tvInput;
     mLooper = looper;
     mTvInputCallback = new TvInputCallback(this);
-    mTvInput->setCallback(mTvInputCallback);
+    mTvInput->setCallback(nullptr);
+    mTvInput->setExtCallback(mTvInputCallback);
 }
 
 JTvInputHal::~JTvInputHal() {
-    mTvInput->setCallback(nullptr);
+    mTvInput->setExtCallback(nullptr);
     JNIEnv* env = AndroidRuntime::getJNIEnv();
     env->DeleteWeakGlobalRef(mThiz);
     mThiz = NULL;
@@ -338,7 +248,21 @@ JTvInputHal* JTvInputHal::createInstance(JNIEnv* env, jobject thiz, const sp<Loo
     return new JTvInputHal(env, thiz, tvInput, looper);
 }
 
+int JTvInputHal::privCmdFromApp(const PrivAppCmdInfo& cmdInfo) {
+    if (mTvInput->privCmdFromApp(cmdInfo) != Result::OK) {
+        ALOGE("Couldn't %s", __FUNCTION__);
+        return BAD_VALUE;
+    }
+    ALOGD("%s end.", __FUNCTION__);
+    return NO_ERROR;
+}
+
 int JTvInputHal::addOrUpdateStream(int deviceId, int streamId, const sp<Surface>& surface) {
+    char c_prop_value[PROPERTY_VALUE_MAX] = {0};
+    property_get("tvinput.hdmiin.buff_type", c_prop_value, "0");
+    int prop_value = (int)atoi(c_prop_value);
+    ALOGW("%s deviceId=%d, streamId=%d, prop=%d", __FUNCTION__, deviceId, streamId, prop_value);
+
     Mutex::Autolock autoLock(&mStreamLock);
     KeyedVector<int, Connection>& connections = mConnections.editValueFor(deviceId);
     if (connections.indexOfKey(streamId) < 0) {
@@ -349,20 +273,40 @@ int JTvInputHal::addOrUpdateStream(int deviceId, int streamId, const sp<Surface>
         // Nothing to do
         return NO_ERROR;
     }
+    connection.mStreamType = (prop_value == 1) ? TV_STREAM_TYPE_BUFFER_PRODUCER : TV_STREAM_TYPE_INDEPENDENT_VIDEO_SOURCE;
     // Clear the surface in the connection.
     if (connection.mSurface != NULL) {
         if (connection.mStreamType == TV_STREAM_TYPE_INDEPENDENT_VIDEO_SOURCE) {
             if (Surface::isValid(connection.mSurface)) {
-                connection.mSurface->setSidebandStream(NULL);
+                if (connection.mCancelSourceHandle != NULL) {
+                    ALOGW("%s setSidebandStream cacnelBuffer", __FUNCTION__);
+                    connection.mSurface->setSidebandStream(connection.mCancelSourceHandle);
+                    connection.mCancelSourceHandle.clear();
+                    connection.mCancelSourceHandle = NULL;
+                } else {
+                    ALOGW("%s setSidebandStream NULL because connection.mSurface invalid", __FUNCTION__);
+                    connection.mSurface->setSidebandStream(NULL);
+                }
             }
         }
         connection.mSurface.clear();
     }
-    if (connection.mSourceHandle == NULL && connection.mThread == NULL) {
+    int32_t extInfo = 2;
+    bool bSidebandFlow = connection.mStreamType == TV_STREAM_TYPE_INDEPENDENT_VIDEO_SOURCE ? true : false;
+    tv_stream_config_ext_t streamConfig;
+    streamConfig.device_id = deviceId;
+    streamConfig.base_config.stream_id = streamId;
+    if (!bSidebandFlow) {
+        streamConfig.base_config.type = TV_STREAM_TYPE_BUFFER_PRODUCER;
+    } else {
+        extInfo = 1;
+        streamConfig.base_config.type = TV_STREAM_TYPE_INDEPENDENT_VIDEO_SOURCE;
+    }
+    if (connection.mSourceHandle == NULL) {
         // Need to configure stream
         Result result = Result::UNKNOWN;
         hidl_vec<TvStreamConfig> list;
-        mTvInput->getStreamConfigurations(deviceId,
+        mTvInput->getStreamConfigurations_ext(deviceId,
                 [&result, &list](Result res, hidl_vec<TvStreamConfig> configs) {
                     result = res;
                     if (res == Result::OK) {
@@ -375,7 +319,7 @@ int JTvInputHal::addOrUpdateStream(int deviceId, int streamId, const sp<Surface>
         }
         int configIndex = -1;
         for (size_t i = 0; i < list.size(); ++i) {
-            if (list[i].streamId == streamId) {
+            if (list[i].base.streamId == streamId) {
                 configIndex = i;
                 break;
             }
@@ -384,31 +328,79 @@ int JTvInputHal::addOrUpdateStream(int deviceId, int streamId, const sp<Surface>
             ALOGE("Cannot find a config with given stream ID: %d", streamId);
             return BAD_VALUE;
         }
-        connection.mStreamType = TV_STREAM_TYPE_INDEPENDENT_VIDEO_SOURCE;
+
+        Result infores = mTvInput->setPreviewInfo(deviceId, streamId, 0, 0, 0, 0, extInfo);
+        if (infores != Result::OK) {
+            ALOGE("initPreviewBuffPoll setPreviewInfo failed");
+            return BAD_VALUE;
+        }
+        if (!bSidebandFlow) {
+            ALOGW("getStreamConfigurations_ext again after setPreviewInfo");
+            mTvInput->getStreamConfigurations_ext(deviceId,
+                    [&result, &list](Result res, hidl_vec<TvStreamConfig> configs) {
+                        result = res;
+                        if (res == Result::OK) {
+                            list = configs;
+                        }
+                    });
+            if (result != Result::OK) {
+                ALOGE("Couldn't get stream configs for device id:%d result:%d", deviceId, result);
+                return UNKNOWN_ERROR;
+            }
+            streamConfig.width = list[configIndex].width;
+            streamConfig.height = list[configIndex].height;
+            streamConfig.usage = list[configIndex].usage;
+            streamConfig.format = list[configIndex].format;//HAL_PIXEL_FORMAT_YCrCb_NV12;
+            streamConfig.buffCount = list[configIndex].buffCount;
+            ALOGD("stream config info: width=%d, height=%d, format=%d, usage=%" PRIu64", buffCount=%d", streamConfig.width, streamConfig.height, streamConfig.format, streamConfig.usage, streamConfig.buffCount);
+            if (connection.mThread == NULL) {
+                connection.mThread = new BufferProducerThread(this, mTvInput, deviceId);
+            }
+            int buffSize = connection.mThread->initPreviewBuffPoll(surface, &streamConfig);
+            int compareCount = list[configIndex].buffCount + 1;
+            if (buffSize != compareCount) {
+                ALOGE("initPreviewBuffPoll size failed");
+                return BAD_VALUE;
+            }
+        }
 
         result = Result::UNKNOWN;
         const native_handle_t* sidebandStream;
-        mTvInput->openStream(deviceId, streamId,
-                [&result, &sidebandStream](Result res, const native_handle_t* handle) {
+        const native_handle_t* sidebandCancelStream = nullptr;
+        mTvInput->openStream_ext(deviceId, streamId, connection.mStreamType,
+                [&result, &sidebandStream, &sidebandCancelStream](Result res, const native_handle_t* handle, const native_handle_t* cancelHandle) {
                     result = res;
                     if (res == Result::OK) {
                         if (handle) {
                             sidebandStream = native_handle_clone(handle);
-                        } else {
-                            result = Result::UNKNOWN;
+                            ALOGW("receiver sidebandStream handle = %p", handle);
                         }
+                        if (cancelHandle) {
+                            sidebandCancelStream = native_handle_clone(cancelHandle);
+                            ALOGW("receiver sidebandCancelStream handle = %p", cancelHandle);
+                        }
+                        result = Result::OK;
                     }
                 });
-        if (result != Result::OK) {
+        if (result != Result::OK && bSidebandFlow) {
             ALOGE("Couldn't open stream. device id:%d stream id:%d result:%d", deviceId, streamId,
                     result);
             return UNKNOWN_ERROR;
+        } else if (bSidebandFlow) {
+            connection.mSourceHandle = NativeHandle::create((native_handle_t*)sidebandStream, true);
+            if (sidebandCancelStream) {
+                connection.mCancelSourceHandle = NativeHandle::create((native_handle_t*)sidebandCancelStream, true);
+            }
         }
-        connection.mSourceHandle = NativeHandle::create((native_handle_t*)sidebandStream, true);
     }
     connection.mSurface = surface;
-    if (connection.mSurface != nullptr) {
+    if (connection.mSurface != nullptr && bSidebandFlow) {
+        ALOGW("%s prepared setSidebandStream", __FUNCTION__);
         connection.mSurface->setSidebandStream(connection.mSourceHandle);
+    } else if (connection.mSurface != nullptr) {
+        ALOGD("start TvInputBufferProducerThread");
+        connection.mThread->run("TvInputBufferProducerThread");
+        connection.mThread->setSurface(surface);
     }
     return NO_ERROR;
 }
@@ -417,35 +409,55 @@ int JTvInputHal::removeStream(int deviceId, int streamId) {
     Mutex::Autolock autoLock(&mStreamLock);
     KeyedVector<int, Connection>& connections = mConnections.editValueFor(deviceId);
     if (connections.indexOfKey(streamId) < 0) {
+        ALOGW("%s connections.indexOfKey(streamId) < 0 streamId=%d", __FUNCTION__, streamId);
         return BAD_VALUE;
     }
     Connection& connection = connections.editValueFor(streamId);
     if (connection.mSurface == NULL) {
+        ALOGW("%s connection.mSurface == NULL streamId=%d", __FUNCTION__, streamId);
         // Nothing to do
         return NO_ERROR;
     }
     if (Surface::isValid(connection.mSurface)) {
-        connection.mSurface->setSidebandStream(NULL);
-    }
-    connection.mSurface.clear();
-    if (connection.mThread != NULL) {
-        connection.mThread->shutdown();
-        connection.mThread.clear();
+        if (connection.mStreamType == TV_STREAM_TYPE_INDEPENDENT_VIDEO_SOURCE) {
+            if (connection.mCancelSourceHandle != NULL) {
+                ALOGW("%s setSidebandStream cacnelBuffer", __FUNCTION__);
+                connection.mSurface->setSidebandStream(connection.mCancelSourceHandle);
+            } else {
+                ALOGW("%s setSidebandStream NULL because connection.mSurface invalid", __FUNCTION__);
+                connection.mSurface->setSidebandStream(NULL);
+            }
+        }
     }
+    ALOGW("%s start mTvInput->closeStream deviceId=%d, streamId=%d", __FUNCTION__, deviceId, streamId);
     if (mTvInput->closeStream(deviceId, streamId) != Result::OK) {
         ALOGE("Couldn't close stream. device id:%d stream id:%d", deviceId, streamId);
         return BAD_VALUE;
     }
+    if (connection.mThread != NULL) {
+        ALOGW("%s exit connection.mThread", __FUNCTION__);
+        connection.mThread->shutdown();
+        connection.mThread->requestExit();
+        connection.mThread->requestExitAndWait();
+        connection.mThread = NULL;
+    }
+    connection.mSurface.clear();
+    connection.mSurface = NULL;
     if (connection.mSourceHandle != NULL) {
         connection.mSourceHandle.clear();
     }
+    if (connection.mCancelSourceHandle != NULL) {
+        connection.mCancelSourceHandle.clear();
+        connection.mCancelSourceHandle = NULL;
+    }
+    ALOGD("%s end.", __FUNCTION__);
     return NO_ERROR;
 }
 
 const hidl_vec<TvStreamConfig> JTvInputHal::getStreamConfigs(int deviceId) {
     Result result = Result::UNKNOWN;
     hidl_vec<TvStreamConfig> list;
-    mTvInput->getStreamConfigurations(deviceId,
+    mTvInput->getStreamConfigurations_ext(deviceId,
             [&result, &list](Result res, hidl_vec<TvStreamConfig> configs) {
                 result = res;
                 if (res == Result::OK) {
@@ -459,9 +471,11 @@ const hidl_vec<TvStreamConfig> JTvInputHal::getStreamConfigs(int deviceId) {
 }
 
 void JTvInputHal::onDeviceAvailable(const TvInputDeviceInfo& info) {
+    ALOGW("onDeviceAvailable %d", info.base.deviceId);
+
     {
         Mutex::Autolock autoLock(&mLock);
-        mConnections.add(info.deviceId, KeyedVector<int, Connection>());
+        mConnections.add(info.base.deviceId, KeyedVector<int, Connection>());
     }
     JNIEnv* env = AndroidRuntime::getJNIEnv();
 
@@ -469,22 +483,22 @@ void JTvInputHal::onDeviceAvailable(const TvInputDeviceInfo& info) {
             gTvInputHardwareInfoBuilderClassInfo.clazz,
             gTvInputHardwareInfoBuilderClassInfo.constructor);
     env->CallObjectMethod(
-            builder, gTvInputHardwareInfoBuilderClassInfo.deviceId, info.deviceId);
+            builder, gTvInputHardwareInfoBuilderClassInfo.deviceId, info.base.deviceId);
     env->CallObjectMethod(
-            builder, gTvInputHardwareInfoBuilderClassInfo.type, info.type);
-    if (info.type == TvInputType::HDMI) {
+            builder, gTvInputHardwareInfoBuilderClassInfo.type, info.base.type);
+    if (info.base.type == TvInputType::HDMI) {
         env->CallObjectMethod(
-                builder, gTvInputHardwareInfoBuilderClassInfo.hdmiPortId, info.portId);
+                builder, gTvInputHardwareInfoBuilderClassInfo.hdmiPortId, info.base.portId);
     }
     env->CallObjectMethod(
             builder, gTvInputHardwareInfoBuilderClassInfo.cableConnectionStatus,
-            info.cableConnectionStatus);
+            info.base.cableConnectionStatus);
     env->CallObjectMethod(
-            builder, gTvInputHardwareInfoBuilderClassInfo.audioType, info.audioType);
-    if (info.audioType != AudioDevice::NONE) {
-        uint8_t buffer[info.audioAddress.size() + 1];
-        memcpy(buffer, info.audioAddress.data(), info.audioAddress.size());
-        buffer[info.audioAddress.size()] = '\0';
+            builder, gTvInputHardwareInfoBuilderClassInfo.audioType, info.base.audioType);
+    if (info.base.audioType != AudioDevice::NONE) {
+        uint8_t buffer[info.base.audioAddress.size() + 1];
+        memcpy(buffer, info.base.audioAddress.data(), info.base.audioAddress.size());
+        buffer[info.base.audioAddress.size()] = '\0';
         jstring audioAddress = env->NewStringUTF(reinterpret_cast<const char *>(buffer));
         env->CallObjectMethod(
                 builder, gTvInputHardwareInfoBuilderClassInfo.audioAddress, audioAddress);
@@ -503,6 +517,8 @@ void JTvInputHal::onDeviceAvailable(const TvInputDeviceInfo& info) {
 }
 
 void JTvInputHal::onDeviceUnavailable(int deviceId) {
+    ALOGW("onDeviceUnavailable deviceId=%d", deviceId);
+
     {
         Mutex::Autolock autoLock(&mLock);
         KeyedVector<int, Connection>& connections = mConnections.editValueFor(deviceId);
@@ -533,8 +549,7 @@ void JTvInputHal::onStreamConfigurationsChanged(int deviceId, int cableConnectio
                         cableConnectionStatus);
 }
 
-void JTvInputHal::onCaptured(int deviceId, int streamId, uint32_t seq, bool succeeded) {
-    sp<BufferProducerThread> thread;
+void JTvInputHal::onCaptured(int deviceId, int streamId, uint64_t buffId, int buffSeq, buffer_handle_t handle, bool succeeded) {
     {
         Mutex::Autolock autoLock(&mLock);
         KeyedVector<int, Connection>& connections = mConnections.editValueFor(deviceId);
@@ -543,39 +558,389 @@ void JTvInputHal::onCaptured(int deviceId, int streamId, uint32_t seq, bool succ
             ALOGE("capture thread not existing.");
             return;
         }
-        thread = connection.mThread;
-    }
-    thread->onCaptured(seq, succeeded);
-    if (seq == 0) {
-        JNIEnv* env = AndroidRuntime::getJNIEnv();
-        env->CallVoidMethod(
-                mThiz,
-                gTvInputHalClassInfo.firstFrameCaptured,
-                deviceId,
-                streamId);
+        connection.mThread->onCaptured(buffId, buffSeq, handle, succeeded);
     }
+    // if (seq == 0) {
+    //     JNIEnv* env = AndroidRuntime::getJNIEnv();
+    //     env->CallVoidMethod(
+    //             mThiz,
+    //             gTvInputHalClassInfo.firstFrameCaptured,
+    //             deviceId,
+    //             streamId);
+    // }
+}
+
+void JTvInputHal::onPrivCmdToApp(int deviceId, const PrivAppCmdInfo& cmdInfo) {
+    JNIEnv* env = AndroidRuntime::getJNIEnv();
+
+    jclass bundleClass = env->FindClass("android/os/Bundle");
+    jmethodID initMID = env->GetMethodID(bundleClass, "<init>", "()V");
+    jobject jBundleObj = env->NewObject(bundleClass, initMID);
+
+    //TODO
+    /*jmethodID putObjectMID = env->GetMethodID(bundleClass, "putObject", "(Ljava/lang/String;Ljava/lang/Object;)V");
+    if (cmdInfo.data && !cmdInfo.data.empty()) {
+        for (int i=0; i<cmdInfo.data.size(); i++) {
+            env->CallVoidMethod(jBundleObj, putObjectMID, cmdInfo.data[i].key, cmdInfo.data[i].value);
+        }
+    }*/
+    //std::vector<hardware::tv::input::V1_0::PrivAppCmdBundle> cmdData;
+    //cmdData.clear();
+    env->CallVoidMethod(mThiz, gTvInputHalClassInfo.privCmdToApp, deviceId,
+            env->NewStringUTF(cmdInfo.action.c_str()), jBundleObj);
 }
 
-JTvInputHal::NotifyHandler::NotifyHandler(JTvInputHal* hal, const TvInputEvent& event) {
+JTvInputHal::NotifyHandler::NotifyHandler(JTvInputHal* hal, const TvInputEventExt& event) {
     mHal = hal;
     mEvent = event;
 }
 
 void JTvInputHal::NotifyHandler::handleMessage(const Message& message) {
+    if (mEvent.type == TvInputEventType::STREAM_CAPTURE_SUCCEEDED) {
+        //ALOGV("%s in, mEvent.type = %d", __FUNCTION__, mEvent.type);
+    } else {
+        ALOGW("%s in, mEvent.type = %d", __FUNCTION__, mEvent.type);
+    }
     switch (mEvent.type) {
         case TvInputEventType::DEVICE_AVAILABLE: {
             mHal->onDeviceAvailable(mEvent.deviceInfo);
         } break;
         case TvInputEventType::DEVICE_UNAVAILABLE: {
-            mHal->onDeviceUnavailable(mEvent.deviceInfo.deviceId);
+            mHal->onDeviceUnavailable(mEvent.deviceInfo.base.deviceId);
         } break;
         case TvInputEventType::STREAM_CONFIGURATIONS_CHANGED: {
-            int cableConnectionStatus = static_cast<int>(mEvent.deviceInfo.cableConnectionStatus);
-            mHal->onStreamConfigurationsChanged(mEvent.deviceInfo.deviceId, cableConnectionStatus);
+            int cableConnectionStatus = static_cast<int>(mEvent.deviceInfo.base.cableConnectionStatus);
+            mHal->onStreamConfigurationsChanged(mEvent.deviceInfo.base.deviceId, cableConnectionStatus);
+        } break;
+        case TvInputEventType::STREAM_CAPTURE_SUCCEEDED:
+        {
+            mHal->onCaptured(mEvent.deviceInfo.base.deviceId,
+                             mEvent.deviceInfo.streamId,
+                             mEvent.capture_result.buffId,
+                             mEvent.capture_result.buffSeq,
+                             mEvent.capture_result.buffer,
+                             true /* succeeded */);
+
+        } break;
+        case TvInputEventType::STREAM_CAPTURE_FAILED:
+        {
+            mHal->onCaptured(mEvent.deviceInfo.base.deviceId,
+                             mEvent.deviceInfo.streamId,
+                             mEvent.capture_result.buffId,
+                             mEvent.capture_result.buffSeq,
+                             mEvent.capture_result.buffer,
+                             false  /*failed*/ );
+        } break;
+        case TvInputEventType::PRIV_CMD_TO_APP:
+        {
+            mHal->onPrivCmdToApp(mEvent.deviceInfo.base.deviceId,
+                            mEvent.priv_app_cmd);
         } break;
         default:
             ALOGE("Unrecognizable event");
     }
+
+}
+
+JTvInputHal::BufferProducerThread::BufferProducerThread(
+        JTvInputHal* hal, sp<ITvInput> tvinput, int deviceId)
+    : Thread(false),
+      mParentHal(hal),
+      mTvInputPtr(tvinput){
+}
+
+status_t JTvInputHal::BufferProducerThread::readyToRun() {
+      mBuffer = NULL;
+      mBufferState = RELEASED;
+      mSeq = 0;
+      mCurrBuffId = 0;
+      mFirstCaptured = true;
+      mShutdown = false;
+      mAlreadyShowSignal = false;
+      return OK;
+}
+
+status_t JTvInputHal::BufferProducerThread::configPreviewBuff() {
+    sp<ANativeWindow> anw(mSurface);
+    ALOGV("%s in", __FUNCTION__);
+    status_t err = native_window_api_connect(anw.get(), NATIVE_WINDOW_API_CAMERA);
+    if (err != NO_ERROR) {
+        ALOGE("%s native_window_api_connect failed. err=%d", __FUNCTION__, err);
+        return err;
+    } else {
+        ALOGV("native_window_api_connect succeed");
+    }
+
+    int consumerUsage = 0;
+    err = anw.get()->query(anw.get(), NATIVE_WINDOW_CONSUMER_USAGE_BITS, &consumerUsage);
+    if (err != NO_ERROR) {
+        ALOGW("failed to get consumer usage bits. ignoring");
+        err = NO_ERROR;
+    }
+    ALOGD("consumerUsage=%d, usage_project=%d", consumerUsage, (consumerUsage & GRALLOC_USAGE_PROTECTED));
+
+    err = native_window_set_usage(anw.get(), (mStream.usage|consumerUsage));
+    if (err != NO_ERROR) {
+        ALOGE("%s native_window_set_usage failed.", __FUNCTION__);
+        return err;
+    }
+
+    // dequeueBuffer cannot time out
+    mSurface->setDequeueTimeout(-1);
+
+    err = native_window_set_buffers_dimensions(
+            anw.get(), mStream.width, mStream.height);
+    if (err != NO_ERROR) {
+        ALOGE("%s native_window_set_buffers_dimensions failed.", __FUNCTION__);
+        return err;
+    }
+    err = native_window_set_buffers_format(anw.get(), 0x15/*mStream.format*/);
+    if (err != NO_ERROR) {
+        ALOGE("%s native_window_set_buffers_format failed.", __FUNCTION__);
+        return err;
+    }
+    err = native_window_set_scaling_mode(anw.get(), NATIVE_WINDOW_SCALING_MODE_SCALE_TO_WINDOW);
+    if (err != NO_ERROR) {
+        ALOGE("%s native_window_set_scaling_mode failed.", __FUNCTION__);
+        return err;
+    }
+    int transform = 0;
+    err = native_window_set_buffers_transform(anw.get(), transform);
+    if (err != NO_ERROR) {
+        ALOGE("%s native_window_set_buffers_transform failed.", __FUNCTION__);
+        return err;
+    }
+
+    int minUndequeuedBufs = 0;
+    err = anw.get()->query(anw.get(),
+            NATIVE_WINDOW_MIN_UNDEQUEUED_BUFFERS, &minUndequeuedBufs);
+    if (err != NO_ERROR) {
+        ALOGE("error pushing blank frames: MIN_UNDEQUEUED_BUFFERS query "
+                "failed: %s (%d)", strerror(-err), -err);
+        return err;
+    }
+    ALOGD("minUndequeuedBufs = %d", minUndequeuedBufs);
+    err = native_window_set_buffer_count(anw.get(), mStream.buffCount + 1);
+    if (err != NO_ERROR) {
+        ALOGE("%s native_window_set_buffer_count failed.", __FUNCTION__);
+        return err;
+    }
+    return NO_ERROR;
+}
+
+int JTvInputHal::BufferProducerThread::initPreviewBuffPoll(const sp<Surface>& surface, const tv_stream_config_ext_t* stream)
+{
+    Mutex::Autolock autoLock(&mLock);
+    ALOGD("%s in", __FUNCTION__);
+    status_t err = NO_ERROR;
+
+    mSurface = surface;
+    mDeviceId = stream->device_id;
+    memcpy(&mStream, stream, sizeof(mStream));
+
+    if(configPreviewBuff() != NO_ERROR) {
+        ALOGE("failed configPreviewBuff, please check your config.");
+        return -1;
+    }
+
+    sp<ANativeWindow> anw(mSurface);
+    int buffCount = mStream.buffCount + 1;
+    mTvHalPreviewBuff.resize(buffCount);
+    mParentHal->mPreviewBuffer.resize(buffCount);
+    static_cast<Surface*>(anw.get())->getIGraphicBufferProducer()->allowAllocation(true);
+    for (int i=0; i<buffCount; i++) {
+        ANativeWindowBuffer_t* buffer = NULL;
+        err = native_window_dequeue_buffer_and_wait(anw.get(), &buffer);
+        // err = anw->dequeueBuffer(anw.get(), &buffer, &mTvHalPreviewBuff[i].bufferFenceFd);
+        if (err != NO_ERROR) {
+            ALOGE("error %d while dequeueing buffer to surface", err);
+            return -1;
+        }
+        sp<GraphicBuffer> gbPtr(GraphicBuffer::from(buffer));
+        ALOGD("%s handle = %p, buffId = %" PRIu64, __FUNCTION__, buffer->handle, gbPtr->getId());
+        mTvHalPreviewBuff[i].buffId = gbPtr->getId();
+        mTvHalPreviewBuff[i].buffStatus = BUFF_STATUS_DEQUEUED;
+        mTvHalPreviewBuff[i].mGraphicBuffer = gbPtr;
+        mTvHalPreviewBuff[i].anwbPtr = buffer;
+        mParentHal->mPreviewBuffer[i].bufferId = gbPtr->getId();
+        mParentHal->mPreviewBuffer[i].buffer = buffer->handle;
+        if (i == buffCount - 1) {
+            ALOGW("%s index=%d, signal handle = %p, buffId = %" PRIu64,
+                __FUNCTION__, i, buffer->handle, gbPtr->getId());
+        }
+        mTvInputPtr->setSinglePreviewBuffer(mParentHal->mPreviewBuffer[i]);
+        buffer = NULL;
+        gbPtr = NULL;
+    }
+    static_cast<Surface*>(anw.get())->getIGraphicBufferProducer()->allowAllocation(false);
+    return mTvHalPreviewBuff.size();
+}
+
+void JTvInputHal::BufferProducerThread::setSurface(const sp<Surface>& surface) {
+    Mutex::Autolock autoLock(&mLock);
+    ALOGD("%s in", __FUNCTION__);
+    mCondition.broadcast();
+}
+
+void JTvInputHal::BufferProducerThread::surfaceBuffDestory() {
+    status_t err = NO_ERROR;
+    sp<ANativeWindow> anw(mSurface);
+    if (!mTvHalPreviewBuff.empty()) {
+        for (int i=0; i< mTvHalPreviewBuff.size(); i++) {
+            ALOGD("buff[%d], bufferId[%" PRIu64"] buffStatus = %d", i, mTvHalPreviewBuff[i].buffId, mTvHalPreviewBuff[i].buffStatus);
+            if (mTvHalPreviewBuff[i].buffStatus == BUFF_STATUS_DEQUEUED) {
+                err = anw->cancelBuffer(anw.get(), mTvHalPreviewBuff[i].mGraphicBuffer.get(), -1);
+                if (err != NO_ERROR) {
+                    ALOGE("%s cancelBuffer failed.", __FUNCTION__);
+                }
+            }
+            mTvHalPreviewBuff[i].mGraphicBuffer.clear();
+            mTvHalPreviewBuff[i].mGraphicBuffer = NULL;
+            mTvHalPreviewBuff[i].anwbPtr.clear();
+            mTvHalPreviewBuff[i].anwbPtr = NULL;
+
+            mParentHal->mPreviewBuffer[i].buffer = NULL;
+        }
+        mTvHalPreviewBuff.clear();
+        mParentHal->mPreviewBuffer.clear();
+    }
+    mBuffer.clear();
+    mBuffer = NULL;
+    err = native_window_api_disconnect(anw.get(), NATIVE_WINDOW_API_CAMERA);
+    if (err != NO_ERROR) {
+        ALOGE("%s native_window_api_disconnect failed.", __FUNCTION__);
+    } else {
+        ALOGD("%s native_window_api_disconnect succeeful.", __FUNCTION__);
+    }
+    mSurface.clear();
+    mSurface = NULL;
+}
+
+void JTvInputHal::BufferProducerThread::onCaptured(uint64_t buffId, int buffSeq, buffer_handle_t handle, bool succeeded) {
+    status_t err = NO_ERROR;
+    if (mShutdown) {
+        return;
+    }
+    sp<ANativeWindow> anw(mSurface);
+    if (buffSeq != mSeq) {
+        ALOGW("Incorrect sequence value: expected %u actual %u", mSeq, buffSeq);
+        mSeq = buffSeq;
+    }
+    // mCondition.broadcast();
+    if (succeeded && anw != NULL) {
+        //ALOGV("%s buffSeq=%d, buffId = %" PRIu64, __FUNCTION__, buffSeq, buffId);
+        if (0 == buffId && !mAlreadyShowSignal) {
+            mAlreadyShowSignal = true;
+            ALOGW("%s show signal", __FUNCTION__);
+            err = anw->queueBuffer(anw.get(), mTvHalPreviewBuff[mTvHalPreviewBuff.size()-1].mGraphicBuffer.get(), -1);
+            return;
+        }
+        if (mAlreadyShowSignal) {
+            ALOGW("%s mAlreadyShowSignal", __FUNCTION__);
+            return;
+        }
+        for (int i=0; i<mTvHalPreviewBuff.size(); i++) {
+            if (buffId == mTvHalPreviewBuff[i].buffId) {
+                err = anw->queueBuffer(anw.get(), mTvHalPreviewBuff[i].mGraphicBuffer.get(), -1);
+                if (err != NO_ERROR) {
+                    ALOGE("error %d while queueing buffer to surface", err);
+                    return;
+                } else {
+                    //ALOGV("queueBuffer succeed mFirstCaptured=%d buff i=%d id = %" PRIu64,
+                    //    mFirstCaptured, i, mTvHalPreviewBuff[i].buffId);
+                    mTvHalPreviewBuff[i].buffStatus = BUFF_STATUS_QUEUED;
+                    mCurrBuffId = buffId;
+                    if (!mFirstCaptured) {
+                        mBufferState = RELEASED;
+                        mCondition.broadcast();
+                    } else {
+                        mFirstCaptured = false;
+                        mBufferState = CAPTURING;
+                    }
+                    break;
+                }
+            }
+        }
+    }
+}
+
+void JTvInputHal::BufferProducerThread::shutdown() {
+    Mutex::Autolock autoLock(&mLock);
+    mShutdown = true;
+    surfaceBuffDestory();
+}
+
+bool JTvInputHal::BufferProducerThread::threadLoop() {
+    Mutex::Autolock autoLock(&mLock);
+    if (mShutdown) {
+        return false;
+    }
+    status_t err = NO_ERROR;
+    if (mSurface == NULL) {
+        err = mCondition.waitRelative(mLock, s2ns(1));
+        // It's OK to time out here.
+        if (err != NO_ERROR && err != TIMED_OUT) {
+            ALOGE("error %d while wating for non-null surface to be set", err);
+            return false;
+        }
+        return true;
+    }
+    sp<ANativeWindow> anw(mSurface);
+    if ((mBuffer == NULL || mBufferState == RELEASED) && !mShutdown && anw != NULL) {
+        ANativeWindowBuffer_t* buffer = NULL;
+        if (mCurrBuffId != 0) {
+            int index = -1;
+            uint64_t dequeueBuffId = 0;
+            //err = native_window_dequeue_buffer_and_wait(anw.get(), &buffer);
+            int fenceFd = -1;
+            err = anw->dequeueBuffer(anw.get(), &buffer, &fenceFd);
+            if (err != NO_ERROR) {
+                ALOGE("error %d while dequeueing buffer to surface", err);
+                return false;
+            } else {
+                dequeueBuffId = GraphicBuffer::from(buffer)->getId();
+                //ALOGV("%s native_window_dequeue_buffer_and_wait succeeded buffId=%" PRIu64, __FUNCTION__, dequeueBuffId);
+            }
+            for (int i=0; i<mTvHalPreviewBuff.size()-1; i++) {
+                //ALOGV("i=%d, buffId=%" PRIu64 ", buffhandle=%p", i, mTvHalPreviewBuff[i].buffId, mTvHalPreviewBuff[i].mGraphicBuffer->getNativeBuffer());
+                if (dequeueBuffId == mTvHalPreviewBuff[i].buffId) {
+                    index = i;
+                    mBuffer = mTvHalPreviewBuff.at(i).anwbPtr;
+                    mCurrBuffId = mTvHalPreviewBuff.at(i).buffId;
+                    mTvHalPreviewBuff[i].buffStatus = BUFF_STATUS_DEQUEUED;
+                    //ALOGV("find the %d right buff %" PRIu64,i, mCurrBuffId);
+                    break;
+                }
+            }
+            sp<Fence> hwcFence(new Fence(fenceFd));
+            //ALOGE("wait fence start %" PRIu64 , dequeueBuffId);
+            //err = hwcFence->waitForever("dequeueBuffer_tv");
+            err = hwcFence->wait(100);
+            //ALOGE("wait fence end %" PRIu64 , dequeueBuffId);
+            if (err != OK) {
+                ALOGE("Fence wait err fenceFd %d err=%d, dequeueBuffId %" PRIx64, fenceFd, err, dequeueBuffId);
+                //anw->cancelBuffer(anw.get(), buffer, fenceFd);
+                //return false;
+            }
+            if (index == -1) {
+                ALOGE("ERROR hapened, after dequeueing, there is no right buff id");
+                return false;
+            }
+        } else {
+            mBuffer = mTvHalPreviewBuff.front().anwbPtr;
+            mCurrBuffId = mTvHalPreviewBuff.front().buffId;
+        }
+        if (mBuffer == NULL) {
+            ALOGE("no buff dequeue");
+            return false;
+        }
+        //ALOGV("before request: capture hanele=%p mCurrBuffId=%" PRIu64, mBuffer.get()->handle, mCurrBuffId);
+        mBufferState = CAPTURING;
+        mTvInputPtr->requestCapture(mDeviceId, mStream.base_config.stream_id, mCurrBuffId, mBuffer.get()->handle, ++mSeq);
+        buffer = NULL;
+    }
+
+    return true;
 }
 
 JTvInputHal::TvInputCallback::TvInputCallback(JTvInputHal* hal) {
@@ -583,7 +948,12 @@ JTvInputHal::TvInputCallback::TvInputCallback(JTvInputHal* hal) {
 }
 
 Return<void> JTvInputHal::TvInputCallback::notify(const TvInputEvent& event) {
-    mHal->mLooper->sendMessage(new NotifyHandler(mHal, event), static_cast<int>(event.type));
+    //mHal->mLooper->sendMessage(new NotifyHandler(mHal, event), static_cast<int>(event.type));
+    return Void();
+}
+
+Return<void> JTvInputHal::TvInputCallback::notify_ext(const TvInputEventExt& extEvent) {
+    mHal->mLooper->sendMessage(new NotifyHandler(mHal, extEvent), static_cast<int>(extEvent.type));
     return Void();
 }
 
@@ -595,6 +965,53 @@ static jlong nativeOpen(JNIEnv* env, jobject thiz, jobject messageQueueObj) {
     return (jlong)JTvInputHal::createInstance(env, thiz, messageQueue->getLooper());
 }
 
+static int nativePrivCmdFromApp(JNIEnv* env, jclass clazz, jlong ptr, jstring action, jobject data) {
+    JTvInputHal* tvInputHal = (JTvInputHal*)ptr;
+    const char* maction = env->GetStringUTFChars(action, NULL);
+
+    jclass bundleClass = env->FindClass("android/os/Bundle");
+    jmethodID keySetMID = env->GetMethodID(bundleClass, "keySet", "()Ljava/util/Set;");
+    jobject ketSetObj = env->CallObjectMethod(data, keySetMID);
+    jmethodID bundleStrMID = env->GetMethodID(bundleClass, "getString", "(Ljava/lang/String;)Ljava/lang/String;");
+
+    jclass setClass = env->FindClass("java/util/Set");
+    jmethodID iteratorMID = env->GetMethodID(setClass, "iterator", "()Ljava/util/Iterator;");
+    jobject iteratorObj = env->CallObjectMethod(ketSetObj, iteratorMID);
+    jclass iteratorClass = env->FindClass("java/util/Iterator");
+    jmethodID hasNextMID = env->GetMethodID(iteratorClass, "hasNext", "()Z");
+    jmethodID nextMID = env->GetMethodID(iteratorClass, "next", "()Ljava/lang/Object;");
+
+    PrivAppCmdInfo cmdInfo;
+    std::vector<PrivAppCmdBundle> cmdData;
+    cmdData.clear();
+    while (env->CallBooleanMethod(iteratorObj, hasNextMID)) {
+        jstring keyJS = (jstring)env->CallObjectMethod(iteratorObj, nextMID);
+        const char *keyStr = env->GetStringUTFChars(keyJS, NULL);
+        jstring valueJS = (jstring)env->CallObjectMethod(data, bundleStrMID, keyJS);
+        const char *valueStr = env->GetStringUTFChars(valueJS, NULL);
+        PrivAppCmdBundle bundle;
+        bundle.key = keyStr;
+        bundle.value = valueStr;
+        cmdData.push_back(bundle);
+
+        env->ReleaseStringUTFChars(keyJS, keyStr);
+        env->DeleteLocalRef(keyJS);
+        env->ReleaseStringUTFChars(valueJS, valueStr);
+        env->DeleteLocalRef(valueJS);
+    }
+    env->DeleteLocalRef(bundleClass);
+    env->DeleteLocalRef(ketSetObj);
+    env->DeleteLocalRef(setClass);
+    env->DeleteLocalRef(iteratorObj);
+    env->DeleteLocalRef(iteratorClass);
+
+    cmdInfo.action = maction;
+    cmdInfo.data = cmdData;
+    tvInputHal->privCmdFromApp(cmdInfo);
+    env->ReleaseStringUTFChars(action, maction);
+    return 1;
+}
+
 static int nativeAddOrUpdateStream(JNIEnv* env, jclass clazz,
         jlong ptr, jint deviceId, jint streamId, jobject jsurface) {
     JTvInputHal* tvInputHal = (JTvInputHal*)ptr;
@@ -625,14 +1042,14 @@ static jobjectArray nativeGetStreamConfigs(JNIEnv* env, jclass clazz,
                 gTvStreamConfigBuilderClassInfo.clazz,
                 gTvStreamConfigBuilderClassInfo.constructor);
         env->CallObjectMethod(
-                builder, gTvStreamConfigBuilderClassInfo.streamId, configs[i].streamId);
+                builder, gTvStreamConfigBuilderClassInfo.streamId, configs[i].base.streamId);
         env->CallObjectMethod(
                 builder, gTvStreamConfigBuilderClassInfo.type,
                         TV_STREAM_TYPE_INDEPENDENT_VIDEO_SOURCE);
         env->CallObjectMethod(
-                builder, gTvStreamConfigBuilderClassInfo.maxWidth, configs[i].maxVideoWidth);
+                builder, gTvStreamConfigBuilderClassInfo.maxWidth, configs[i].base.maxVideoWidth);
         env->CallObjectMethod(
-                builder, gTvStreamConfigBuilderClassInfo.maxHeight, configs[i].maxVideoHeight);
+                builder, gTvStreamConfigBuilderClassInfo.maxHeight, configs[i].base.maxVideoHeight);
         env->CallObjectMethod(
                 builder, gTvStreamConfigBuilderClassInfo.generation, generation);
 
@@ -663,6 +1080,9 @@ static const JNINativeMethod gTvInputHalMethods[] = {
             (void*) nativeGetStreamConfigs },
     { "nativeClose", "(J)V",
             (void*) nativeClose },
+    { "nativePrivCmdFromApp", "(JLjava/lang/String;Landroid/os/Bundle;)I",
+            (void*) nativePrivCmdFromApp },
+
 };
 
 #define FIND_CLASS(var, className) \
@@ -692,6 +1112,9 @@ int register_android_server_tv_TvInputHal(JNIEnv* env) {
     GET_METHOD_ID(
             gTvInputHalClassInfo.firstFrameCaptured, clazz,
             "firstFrameCapturedFromNative", "(II)V");
+    GET_METHOD_ID(
+                gTvInputHalClassInfo.privCmdToApp, clazz,
+                "privCmdToAppFromNative", "(ILjava/lang/String;Landroid/os/Bundle;)V");
 
     FIND_CLASS(gTvStreamConfigClassInfo.clazz, "android/media/tv/TvStreamConfig");
     gTvStreamConfigClassInfo.clazz = jclass(env->NewGlobalRef(gTvStreamConfigClassInfo.clazz));
