From c9901950d0d9af7b61414348ecddb8989e987230 Mon Sep 17 00:00:00 2001
From: Vaciller <wyc_19533480830@outlook.com>
Date: Fri, 10 Jul 2026 17:24:19 +0800
Subject: [PATCH] feat: add RkAudioSetting integration and HDMI EDID format
 update

Change-Id: I80f60eac899316e3466c2eb80ce548021b8360b8

diff --git a/media/java/android/media/AudioDeviceInfo.java b/media/java/android/media/AudioDeviceInfo.java
index 552a42300839..7a036b04660b 100644
--- a/media/java/android/media/AudioDeviceInfo.java
+++ b/media/java/android/media/AudioDeviceInfo.java
@@ -624,6 +624,8 @@ public final class AudioDeviceInfo {
         INT_TO_EXT_DEVICE_MAPPING.put(AudioSystem.DEVICE_OUT_BLUETOOTH_A2DP_HEADPHONES, TYPE_BLUETOOTH_A2DP);
         INT_TO_EXT_DEVICE_MAPPING.put(AudioSystem.DEVICE_OUT_BLUETOOTH_A2DP_SPEAKER, TYPE_BLUETOOTH_A2DP);
         INT_TO_EXT_DEVICE_MAPPING.put(AudioSystem.DEVICE_OUT_HDMI, TYPE_HDMI);
+        INT_TO_EXT_DEVICE_MAPPING.put(AudioSystem.DEVICE_OUT_HDMI_1, TYPE_HDMI);
+        INT_TO_EXT_DEVICE_MAPPING.put(AudioSystem.DEVICE_OUT_ANLG_DOCK_HEADSET, TYPE_DOCK);
         INT_TO_EXT_DEVICE_MAPPING.put(AudioSystem.DEVICE_OUT_DGTL_DOCK_HEADSET, TYPE_DOCK);
         INT_TO_EXT_DEVICE_MAPPING.put(AudioSystem.DEVICE_OUT_USB_ACCESSORY, TYPE_USB_ACCESSORY);
         INT_TO_EXT_DEVICE_MAPPING.put(AudioSystem.DEVICE_OUT_USB_DEVICE, TYPE_USB_DEVICE);
@@ -633,6 +635,7 @@ public final class AudioDeviceInfo {
         INT_TO_EXT_DEVICE_MAPPING.put(AudioSystem.DEVICE_OUT_HDMI_ARC, TYPE_HDMI_ARC);
         INT_TO_EXT_DEVICE_MAPPING.put(AudioSystem.DEVICE_OUT_HDMI_EARC, TYPE_HDMI_EARC);
         INT_TO_EXT_DEVICE_MAPPING.put(AudioSystem.DEVICE_OUT_SPDIF, TYPE_LINE_DIGITAL);
+        INT_TO_EXT_DEVICE_MAPPING.put(AudioSystem.DEVICE_OUT_SPDIF_1, TYPE_LINE_DIGITAL);
         INT_TO_EXT_DEVICE_MAPPING.put(AudioSystem.DEVICE_OUT_FM, TYPE_FM);
         INT_TO_EXT_DEVICE_MAPPING.put(AudioSystem.DEVICE_OUT_AUX_LINE, TYPE_AUX_LINE);
         INT_TO_EXT_DEVICE_MAPPING.put(AudioSystem.DEVICE_OUT_IP, TYPE_IP);
@@ -649,6 +652,7 @@ public final class AudioDeviceInfo {
         INT_TO_EXT_DEVICE_MAPPING.put(AudioSystem.DEVICE_IN_BLUETOOTH_SCO_HEADSET, TYPE_BLUETOOTH_SCO);
         INT_TO_EXT_DEVICE_MAPPING.put(AudioSystem.DEVICE_IN_WIRED_HEADSET, TYPE_WIRED_HEADSET);
         INT_TO_EXT_DEVICE_MAPPING.put(AudioSystem.DEVICE_IN_HDMI, TYPE_HDMI);
+        INT_TO_EXT_DEVICE_MAPPING.put(AudioSystem.DEVICE_IN_HDMI_1, TYPE_HDMI);
         INT_TO_EXT_DEVICE_MAPPING.put(AudioSystem.DEVICE_IN_TELEPHONY_RX, TYPE_TELEPHONY);
         INT_TO_EXT_DEVICE_MAPPING.put(AudioSystem.DEVICE_IN_BACK_MIC, TYPE_BUILTIN_MIC);
         INT_TO_EXT_DEVICE_MAPPING.put(AudioSystem.DEVICE_IN_DGTL_DOCK_HEADSET, TYPE_DOCK);
@@ -677,9 +681,11 @@ public final class AudioDeviceInfo {
         EXT_TO_INT_DEVICE_MAPPING.put(TYPE_WIRED_HEADPHONES, AudioSystem.DEVICE_OUT_WIRED_HEADPHONE);
         EXT_TO_INT_DEVICE_MAPPING.put(TYPE_LINE_ANALOG, AudioSystem.DEVICE_OUT_LINE);
         EXT_TO_INT_DEVICE_MAPPING.put(TYPE_LINE_DIGITAL, AudioSystem.DEVICE_OUT_SPDIF);
+        EXT_TO_INT_DEVICE_MAPPING.put(TYPE_LINE_DIGITAL, AudioSystem.DEVICE_OUT_SPDIF_1);
         EXT_TO_INT_DEVICE_MAPPING.put(TYPE_BLUETOOTH_SCO, AudioSystem.DEVICE_OUT_BLUETOOTH_SCO);
         EXT_TO_INT_DEVICE_MAPPING.put(TYPE_BLUETOOTH_A2DP, AudioSystem.DEVICE_OUT_BLUETOOTH_A2DP);
         EXT_TO_INT_DEVICE_MAPPING.put(TYPE_HDMI, AudioSystem.DEVICE_OUT_HDMI);
+        EXT_TO_INT_DEVICE_MAPPING.put(TYPE_HDMI, AudioSystem.DEVICE_OUT_HDMI_1);
         EXT_TO_INT_DEVICE_MAPPING.put(TYPE_HDMI_ARC, AudioSystem.DEVICE_OUT_HDMI_ARC);
         EXT_TO_INT_DEVICE_MAPPING.put(TYPE_HDMI_EARC, AudioSystem.DEVICE_OUT_HDMI_EARC);
         EXT_TO_INT_DEVICE_MAPPING.put(TYPE_USB_DEVICE, AudioSystem.DEVICE_OUT_USB_DEVICE);
@@ -707,6 +713,7 @@ public final class AudioDeviceInfo {
         EXT_TO_INT_INPUT_DEVICE_MAPPING.put(
                 TYPE_WIRED_HEADSET, AudioSystem.DEVICE_IN_WIRED_HEADSET);
         EXT_TO_INT_INPUT_DEVICE_MAPPING.put(TYPE_HDMI, AudioSystem.DEVICE_IN_HDMI);
+        EXT_TO_INT_INPUT_DEVICE_MAPPING.put(TYPE_HDMI, AudioSystem.DEVICE_IN_HDMI_1);
         EXT_TO_INT_INPUT_DEVICE_MAPPING.put(TYPE_TELEPHONY, AudioSystem.DEVICE_IN_TELEPHONY_RX);
         EXT_TO_INT_INPUT_DEVICE_MAPPING.put(TYPE_DOCK, AudioSystem.DEVICE_IN_DGTL_DOCK_HEADSET);
         EXT_TO_INT_INPUT_DEVICE_MAPPING.put(
diff --git a/media/java/android/media/AudioManager.java b/media/java/android/media/AudioManager.java
index 67efe3a55db7..2bdf728bbeb4 100644
--- a/media/java/android/media/AudioManager.java
+++ b/media/java/android/media/AudioManager.java
@@ -5561,6 +5561,10 @@ public class AudioManager {
      * The audio output device code for HDMI */
     @UnsupportedAppUsage
     public static final int DEVICE_OUT_HDMI = AudioSystem.DEVICE_OUT_HDMI;
+    /** @hide
+     * The audio output device code for another HDMI */
+    @UnsupportedAppUsage
+    public static final int DEVICE_OUT_HDMI_1 = AudioSystem.DEVICE_OUT_HDMI_1;
     /** @hide
      * The audio output device code for an analog wired headset attached via a
      *  docking station
@@ -5607,6 +5611,10 @@ public class AudioManager {
      * The audio output device code for S/PDIF digital connection.
      */
     public static final int DEVICE_OUT_SPDIF = AudioSystem.DEVICE_OUT_SPDIF;
+    /** @hide
+     * The audio output device code for another S/PDIF digital connection.
+     */
+    public static final int DEVICE_OUT_SPDIF_1 = AudioSystem.DEVICE_OUT_SPDIF_1;
     /** @hide
      * The audio output device code for built-in FM transmitter.
      */
@@ -5653,6 +5661,11 @@ public class AudioManager {
      */
     public static final int DEVICE_IN_HDMI =
                                     AudioSystem.DEVICE_IN_HDMI;
+    /** @hide
+     * The audio input device code for another HDMI
+     */
+    public static final int DEVICE_IN_HDMI_1 =
+                                    AudioSystem.DEVICE_IN_HDMI_1;
     /** @hide
      * The audio input device code for HDMI ARC
      */
diff --git a/media/java/android/media/AudioSystem.java b/media/java/android/media/AudioSystem.java
index 7ccbe51bbd49..0dfc97b936d3 100644
--- a/media/java/android/media/AudioSystem.java
+++ b/media/java/android/media/AudioSystem.java
@@ -280,6 +280,10 @@ public class AudioSystem
     @Retention(RetentionPolicy.SOURCE)
     public @interface DeviceType {}
 
+    // add AUDIO_FORMAT_IEC61937 define
+    /** @hide */
+    public static final int AUDIO_FORMAT_IEC61937       = 0x0D000000;
+
     /**
      * @hide
      * Convert audio format enum values to Bluetooth codec values
@@ -988,6 +992,8 @@ public class AudioSystem
     /** @hide */
     public static final int DEVICE_OUT_HDMI = DEVICE_OUT_AUX_DIGITAL;
     /** @hide */
+    public static final int DEVICE_OUT_HDMI_1 = 0x401;
+    /** @hide */
     @UnsupportedAppUsage
     public static final int DEVICE_OUT_ANLG_DOCK_HEADSET = 0x800;
     /** @hide */
@@ -1014,6 +1020,8 @@ public class AudioSystem
     /** @hide */
     public static final int DEVICE_OUT_SPDIF = 0x80000;
     /** @hide */
+    public static final int DEVICE_OUT_SPDIF_1 = 0x80001;
+    /** @hide */
     @UnsupportedAppUsage
     public static final int DEVICE_OUT_FM = 0x100000;
     /** @hide */
@@ -1077,6 +1085,7 @@ public class AudioSystem
         DEVICE_OUT_ALL_SET.add(DEVICE_OUT_BLUETOOTH_A2DP_HEADPHONES);
         DEVICE_OUT_ALL_SET.add(DEVICE_OUT_BLUETOOTH_A2DP_SPEAKER);
         DEVICE_OUT_ALL_SET.add(DEVICE_OUT_HDMI);
+        DEVICE_OUT_ALL_SET.add(DEVICE_OUT_HDMI_1);
         DEVICE_OUT_ALL_SET.add(DEVICE_OUT_ANLG_DOCK_HEADSET);
         DEVICE_OUT_ALL_SET.add(DEVICE_OUT_DGTL_DOCK_HEADSET);
         DEVICE_OUT_ALL_SET.add(DEVICE_OUT_USB_ACCESSORY);
@@ -1087,6 +1096,7 @@ public class AudioSystem
         DEVICE_OUT_ALL_SET.add(DEVICE_OUT_HDMI_ARC);
         DEVICE_OUT_ALL_SET.add(DEVICE_OUT_HDMI_EARC);
         DEVICE_OUT_ALL_SET.add(DEVICE_OUT_SPDIF);
+        DEVICE_OUT_ALL_SET.add(DEVICE_OUT_SPDIF_1);
         DEVICE_OUT_ALL_SET.add(DEVICE_OUT_FM);
         DEVICE_OUT_ALL_SET.add(DEVICE_OUT_AUX_LINE);
         DEVICE_OUT_ALL_SET.add(DEVICE_OUT_SPEAKER_SAFE);
@@ -1154,6 +1164,8 @@ public class AudioSystem
     /** @hide */
     public static final int DEVICE_IN_HDMI = DEVICE_IN_AUX_DIGITAL;
     /** @hide */
+    public static final int DEVICE_IN_HDMI_1 = DEVICE_BIT_IN | 0x21;
+    /** @hide */
     @UnsupportedAppUsage
     public static final int DEVICE_IN_VOICE_CALL = DEVICE_BIT_IN | 0x40;
     /** @hide */
@@ -1225,6 +1237,7 @@ public class AudioSystem
         DEVICE_IN_ALL_SET.add(DEVICE_IN_BLUETOOTH_SCO_HEADSET);
         DEVICE_IN_ALL_SET.add(DEVICE_IN_WIRED_HEADSET);
         DEVICE_IN_ALL_SET.add(DEVICE_IN_HDMI);
+        DEVICE_IN_ALL_SET.add(DEVICE_IN_HDMI_1);
         DEVICE_IN_ALL_SET.add(DEVICE_IN_TELEPHONY_RX);
         DEVICE_IN_ALL_SET.add(DEVICE_IN_BACK_MIC);
         DEVICE_IN_ALL_SET.add(DEVICE_IN_REMOTE_SUBMIX);
@@ -1292,6 +1305,7 @@ public class AudioSystem
     /** @hide */ public static final String DEVICE_OUT_BLUETOOTH_A2DP_SPEAKER_NAME = "bt_a2dp_spk";
     /** @hide */ public static final String DEVICE_OUT_AUX_DIGITAL_NAME = "aux_digital";
     /** @hide */ public static final String DEVICE_OUT_HDMI_NAME = "hdmi";
+    /** @hide */ public static final String DEVICE_OUT_HDMI1_NAME = "hdmi1";
     /** @hide */ public static final String DEVICE_OUT_ANLG_DOCK_HEADSET_NAME = "analog_dock";
     /** @hide */ public static final String DEVICE_OUT_DGTL_DOCK_HEADSET_NAME = "digital_dock";
     /** @hide */ public static final String DEVICE_OUT_USB_ACCESSORY_NAME = "usb_accessory";
@@ -1302,6 +1316,7 @@ public class AudioSystem
     /** @hide */ public static final String DEVICE_OUT_HDMI_ARC_NAME = "hdmi_arc";
     /** @hide */ public static final String DEVICE_OUT_HDMI_EARC_NAME = "hdmi_earc";
     /** @hide */ public static final String DEVICE_OUT_SPDIF_NAME = "spdif";
+    /** @hide */ public static final String DEVICE_OUT_SPDIF1_NAME = "spdif1";
     /** @hide */ public static final String DEVICE_OUT_FM_NAME = "fm_transmitter";
     /** @hide */ public static final String DEVICE_OUT_AUX_LINE_NAME = "aux_line";
     /** @hide */ public static final String DEVICE_OUT_SPEAKER_SAFE_NAME = "speaker_safe";
@@ -1321,6 +1336,7 @@ public class AudioSystem
     /** @hide */ public static final String DEVICE_IN_BLUETOOTH_SCO_HEADSET_NAME = "bt_sco_hs";
     /** @hide */ public static final String DEVICE_IN_WIRED_HEADSET_NAME = "headset";
     /** @hide */ public static final String DEVICE_IN_AUX_DIGITAL_NAME = "aux_digital";
+    /** @hide */ public static final String DEVICE_IN_HDMI1_NAME = "hdmiin1";
     /** @hide */ public static final String DEVICE_IN_TELEPHONY_RX_NAME = "telephony_rx";
     /** @hide */ public static final String DEVICE_IN_BACK_MIC_NAME = "back_mic";
     /** @hide */ public static final String DEVICE_IN_REMOTE_SUBMIX_NAME = "remote_submix";
@@ -1371,6 +1387,8 @@ public class AudioSystem
             return DEVICE_OUT_BLUETOOTH_A2DP_SPEAKER_NAME;
         case DEVICE_OUT_HDMI:
             return DEVICE_OUT_HDMI_NAME;
+        case DEVICE_OUT_HDMI_1:
+            return DEVICE_OUT_HDMI1_NAME;
         case DEVICE_OUT_ANLG_DOCK_HEADSET:
             return DEVICE_OUT_ANLG_DOCK_HEADSET_NAME;
         case DEVICE_OUT_DGTL_DOCK_HEADSET:
@@ -1391,6 +1409,8 @@ public class AudioSystem
             return DEVICE_OUT_HDMI_EARC_NAME;
         case DEVICE_OUT_SPDIF:
             return DEVICE_OUT_SPDIF_NAME;
+        case DEVICE_OUT_SPDIF_1:
+            return DEVICE_OUT_SPDIF1_NAME;
         case DEVICE_OUT_FM:
             return DEVICE_OUT_FM_NAME;
         case DEVICE_OUT_AUX_LINE:
@@ -2409,4 +2429,3 @@ public class AudioSystem
      */
     final static int NATIVE_EVENT_ROUTING_CHANGE = 1000;
 }
-
diff --git a/services/core/java/com/android/server/audio/AudioService.java b/services/core/java/com/android/server/audio/AudioService.java
index 3ddb0ae96a22..8e0258cb1925 100644
--- a/services/core/java/com/android/server/audio/AudioService.java
+++ b/services/core/java/com/android/server/audio/AudioService.java
@@ -656,7 +656,7 @@ public class AudioService extends IAudioService.Stub
     private ForceControlStreamClient mForceControlStreamClient = null;
     // Used to play ringtones outside system_server
     private volatile IRingtonePlayer mRingtonePlayer;
-
+    private RkAudioSetting mRkAudioSetting = null;
     // Devices for which the volume is fixed (volume is either max or muted)
     Set<Integer> mFixedVolumeDevices = new HashSet<>(Arrays.asList(
             AudioSystem.DEVICE_OUT_ANLG_DOCK_HEADSET,
@@ -666,7 +666,13 @@ public class AudioService extends IAudioService.Stub
             AudioSystem.DEVICE_OUT_HDMI_ARC,
             AudioSystem.DEVICE_OUT_HDMI_EARC
     ));
-
+    // Devices for which the volumes are being ajusted at the same time
+    Set<Integer> mSyncAjustVolumeDevices = new HashSet<>(Arrays.asList(
+        AudioSystem.DEVICE_OUT_HDMI,
+        AudioSystem.DEVICE_OUT_HDMI_1,
+        AudioSystem.DEVICE_OUT_SPDIF,
+        AudioSystem.DEVICE_OUT_SPDIF_1,
+        AudioSystem.DEVICE_OUT_SPEAKER));
     // Devices where the framework sends a full scale audio signal, and controls the volume of
     // the external audio system separately.
     Map<Integer, AbsoluteVolumeDeviceInfo> mAbsoluteVolumeDeviceInfoMap = new ArrayMap<>();
@@ -1003,6 +1009,10 @@ public class AudioService extends IAudioService.Stub
         mSystemServer = systemServer;
         mSettings = settings;
 
+        if (mContext.getPackageManager().hasSystemFeature(PackageManager.FEATURE_ROCKCHIP_AUDIO)) {
+            mRkAudioSetting = new RkAudioSetting(context);
+        }
+
         mPlatformType = AudioSystem.getPlatformType(context);
 
         mIsSingleVolume = AudioSystem.isSingleVolume(context);
@@ -4659,6 +4669,17 @@ public class AudioService extends IAudioService.Stub
         }
     }
 
+    /**
+     * @hide
+     */
+    private boolean isTablet() {
+        String product = SystemProperties.get("ro.target.product","");
+        if(product.equals("tablet")){
+            return true;
+        }
+        return false;
+    }
+
     private class RmtSbmxFullVolDeathHandler implements IBinder.DeathRecipient {
         private IBinder mICallback; // To be notified of client's death
 
@@ -7277,6 +7298,11 @@ public class AudioService extends IAudioService.Stub
                 .set(MediaMetrics.Property.STATE,
                         state == CONNECTION_STATE_CONNECTED ? "connected" : "disconnected")
                 .record();
+
+        if ((attributes.getInternalType() == AudioSystem.DEVICE_OUT_AUX_DIGITAL) && (state == CONNECTION_STATE_CONNECTED)) {
+            updataFormatForEdid();
+        }
+
         mDeviceBroker.setWiredDeviceConnectionState(attributes, state, caller);
     }
 
@@ -7294,6 +7320,17 @@ public class AudioService extends IAudioService.Stub
                 /*delay*/ 0);
     }
 
+    public void updataFormatForEdid() {
+        try {
+            if(mRkAudioSetting == null){
+                return;
+            }
+            mRkAudioSetting.updataFormatForEdid();
+        } catch (RemoteException e) {
+            e.printStackTrace();
+        }
+    }
+
     /**
      * @hide
      * The states that can be used with AudioService.setBluetoothHearingAidDeviceConnectionState()
@@ -8221,7 +8258,14 @@ public class AudioService extends IAudioService.Stub
             } else {
                 index = (getIndex(device) + 5)/10;
             }
-            setStreamVolumeIndex(index, device);
+            if (isTablet() && isSyncAjustVolumeDevice(device) && mStreamType == AudioSystem.STREAM_MUSIC) {
+                for (Integer item : mSyncAjustVolumeDevices) {
+                    index = (getIndex(item) + 5)/10;
+                    setStreamVolumeIndex(index, item);
+                }
+            } else {
+                setStreamVolumeIndex(index, device);
+            }
         }
 
         public void applyAllVolumes() {
@@ -8264,20 +8308,41 @@ public class AudioService extends IAudioService.Stub
                     hasModifyAudioSettings);
         }
 
+        //save the volume of all the devices in mSyncAjustVolumeDevices
+        public void restoreSyncDevicesIndex(){
+            if(mStreamType == AudioSystem.STREAM_MUSIC){
+                for (Integer device : mSyncAjustVolumeDevices) {
+                    System.putIntForUser(mContentResolver,
+                          getSettingNameForDevice(device),
+                         (getIndex(device) + 5)/ 10,
+                          UserHandle.USER_CURRENT);
+                }
+            }
+        }
+
         public boolean setIndex(int index, int device, String caller,
                 boolean hasModifyAudioSettings) {
             boolean changed;
             int oldIndex;
             final boolean isCurrentDevice;
+            int deltaIndex;
             synchronized (mSettingsLock) {
                 synchronized (VolumeStreamState.class) {
                     oldIndex = getIndex(device);
+                    deltaIndex = index - oldIndex;
+                    if (isTablet() && isSyncAjustVolumeDevice(device) && mStreamType == AudioSystem.STREAM_MUSIC) {
+                        int index_tmp;
+                        for (Integer item : mSyncAjustVolumeDevices) {
+                            index_tmp = getIndex(item) + deltaIndex;
+                            index_tmp = getValidIndex(index_tmp, hasModifyAudioSettings);
+                            mIndexMap.put(item, index_tmp);
+                        }
+                    }
                     index = getValidIndex(index, hasModifyAudioSettings);
                     if ((mStreamType == AudioSystem.STREAM_SYSTEM_ENFORCED) && mCameraSoundForced) {
                         index = mIndexMax;
                     }
                     mIndexMap.put(device, index);
-
                     changed = oldIndex != index;
                     // Apply change to all streams using this one as alias if:
                     // - the index actually changed OR
@@ -8775,6 +8840,8 @@ public class AudioService extends IAudioService.Stub
                         streamState.getSettingNameForDevice(device),
                         (streamState.getIndex(device) + 5) / 10,
                         UserHandle.USER_CURRENT);
+                if(isTablet())
+                    streamState.restoreSyncDevicesIndex();
             }
         }
 
@@ -10689,6 +10756,7 @@ public class AudioService extends IAudioService.Stub
         pw.print("  mNotifAliasRing="); pw.println(mNotifAliasRing);
         pw.print("  mFixedVolumeDevices="); pw.println(dumpDeviceTypes(mFixedVolumeDevices));
         pw.print("  mFullVolumeDevices="); pw.println(dumpDeviceTypes(mFullVolumeDevices));
+        pw.print("  mSyncAjustVolumeDevices="); pw.println(dumpDeviceTypes(mSyncAjustVolumeDevices));
         pw.print("  mAbsoluteVolumeDevices.keySet()="); pw.println(dumpDeviceTypes(
                 mAbsoluteVolumeDeviceInfoMap.keySet()));
         pw.print("  mExtVolumeController="); pw.println(mExtVolumeController);
@@ -12498,6 +12566,14 @@ public class AudioService extends IAudioService.Stub
         return mFullVolumeDevices.contains(deviceType);
     }
 
+    private boolean isSyncAjustVolumeDevice(int deviceType) {
+        if (deviceType == AudioSystem.DEVICE_OUT_REMOTE_SUBMIX
+                && mRecordMonitor.isLegacyRemoteSubmixActive()) {
+            return false;
+        }
+        return mSyncAjustVolumeDevices.contains(deviceType);
+    }
+
     /**
      * Returns whether the input device uses absolute volume behavior. This is distinct
      * from Bluetooth A2DP absolute volume behavior ({@link #isA2dpAbsoluteVolumeDevice}).
