From 23655d346fa9c93a559583fe8618a69553a6ba2a Mon Sep 17 00:00:00 2001
From: Vaciller <wyc_19533480830@outlook.com>
Date: Fri, 10 Jul 2026 17:24:52 +0800
Subject: [PATCH] feat: add extcon alias support in WiredAccessoryManager

Change-Id: I2b7c06a108f001f06185ee39c5dcf7c6f7a2a2ce

diff --git a/services/core/java/com/android/server/WiredAccessoryManager.java b/services/core/java/com/android/server/WiredAccessoryManager.java
index b271d7ece733..6162ad36009c 100644
--- a/services/core/java/com/android/server/WiredAccessoryManager.java
+++ b/services/core/java/com/android/server/WiredAccessoryManager.java
@@ -35,6 +35,7 @@ import android.util.Log;
 import android.util.Pair;
 import android.util.Slog;
 import android.view.InputDevice;
+import android.os.FileUtils;
 
 import com.android.internal.R;
 import com.android.server.input.InputManagerService;
@@ -63,9 +64,15 @@ final class WiredAccessoryManager implements WiredAccessoryCallbacks {
     private static final int BIT_USB_HEADSET_DGTL = (1 << 3);
     private static final int BIT_HDMI_AUDIO = (1 << 4);
     private static final int BIT_LINEOUT = (1 << 5);
+    private static final int BIT_DP_AUDIO = (1 << 6);
+    private static final int BIT_HDMI_AUDIO_1 = (1 << 7);
+    private static final int BIT_DP_AUDIO_1 = (1 << 8);
+    private static final int BIT_HDMIIN_AUDIO = (1 << 9);
+    private static final int BIT_HDMIIN_AUDIO_1 = (1 << 10);
     private static final int SUPPORTED_HEADSETS = (BIT_HEADSET | BIT_HEADSET_NO_MIC |
-            BIT_USB_HEADSET_ANLG | BIT_USB_HEADSET_DGTL |
-            BIT_HDMI_AUDIO | BIT_LINEOUT);
+            BIT_USB_HEADSET_ANLG | BIT_USB_HEADSET_DGTL | BIT_HDMI_AUDIO | BIT_LINEOUT |
+            BIT_DP_AUDIO | BIT_HDMI_AUDIO_1 | BIT_DP_AUDIO_1 | BIT_HDMIIN_AUDIO |
+            BIT_HDMIIN_AUDIO_1);
 
     private static final String NAME_H2W = "h2w";
     private static final String NAME_USB_AUDIO = "usb_audio";
@@ -122,7 +129,6 @@ final class WiredAccessoryManager implements WiredAccessoryCallbacks {
                     SW_HEADPHONE_INSERT_BIT | SW_MICROPHONE_INSERT_BIT | SW_LINEOUT_INSERT_BIT);
         }
 
-
         if (ExtconUEventObserver.extconExists()) {
             if (mUseDevInputEventForAudioJack) {
                 Log.w(TAG, "Both input event and extcon are used for audio jack,"
@@ -298,7 +304,23 @@ final class WiredAccessoryManager implements WiredAccessoryCallbacks {
             } else if (headset == BIT_USB_HEADSET_DGTL) {
                 outDevice = AudioManager.DEVICE_OUT_DGTL_DOCK_HEADSET;
             } else if (headset == BIT_HDMI_AUDIO) {
+                Slog.d(TAG, "hdmi_0 plug");
                 outDevice = AudioManager.DEVICE_OUT_HDMI;
+            } else if (headset == BIT_HDMI_AUDIO_1) {
+                Slog.d(TAG, "hdmi_1 plug");
+                outDevice = AudioManager.DEVICE_OUT_HDMI_1;
+            } else if (headset == BIT_DP_AUDIO) {
+                Slog.d(TAG, "dp_0 plug");
+                outDevice = AudioManager.DEVICE_OUT_SPDIF;
+            } else if (headset == BIT_DP_AUDIO_1) {
+                Slog.d(TAG, "dp_1 plug");
+                outDevice = AudioManager.DEVICE_OUT_SPDIF_1;
+            } else if (headset == BIT_HDMIIN_AUDIO) {
+                Slog.d(TAG, "hdmiin_0 plug");
+                inDevice = AudioManager.DEVICE_IN_HDMI;
+            } else if (headset == BIT_HDMIIN_AUDIO_1) {
+                Slog.d(TAG, "hdmiin_1 plug");
+                inDevice = AudioManager.DEVICE_IN_HDMI_1;
             } else {
                 Slog.e(TAG, "setDeviceState() invalid headset type: " + headset);
                 return;
@@ -494,7 +516,9 @@ final class WiredAccessoryManager implements WiredAccessoryCallbacks {
                     ExtconInfo.EXTCON_HEADPHONE,
                     ExtconInfo.EXTCON_MICROPHONE,
                     ExtconInfo.EXTCON_HDMI,
+                    ExtconInfo.EXTCON_DP,
                     ExtconInfo.EXTCON_LINE_OUT,
+                    ExtconInfo.EXTCON_VIDEO_IN,
             });
         }
 
@@ -535,11 +559,19 @@ final class WiredAccessoryManager implements WiredAccessoryCallbacks {
                 updateBit(maskAndState, BIT_HEADSET, status, ExtconInfo.EXTCON_MICROPHONE);
             }
             if (extconInfo.hasCableType(ExtconInfo.EXTCON_HDMI)) {
-                updateBit(maskAndState, BIT_HDMI_AUDIO, status, ExtconInfo.EXTCON_HDMI);
+                //updateBit(maskAndState, BIT_HDMI_AUDIO, status, ExtconInfo.EXTCON_HDMI);
             }
             if (extconInfo.hasCableType(ExtconInfo.EXTCON_LINE_OUT)) {
                 updateBit(maskAndState, BIT_LINEOUT, status, ExtconInfo.EXTCON_LINE_OUT);
             }
+
+            updateBit(maskAndState, BIT_HDMI_AUDIO,     status, "hdmi0");
+            updateBit(maskAndState, BIT_HDMI_AUDIO_1,   status, "hdmi1");
+            updateBit(maskAndState, BIT_DP_AUDIO,       status, "dp0");
+            updateBit(maskAndState, BIT_DP_AUDIO_1,     status, "dp1");
+            updateBit(maskAndState, BIT_HDMIIN_AUDIO,   status, "hdmirx0");
+            updateBit(maskAndState, BIT_HDMIIN_AUDIO_1, status, "hdmirx1");
+
             if (LOG) Slog.v(TAG, "mask " + maskAndState[0] + " state " + maskAndState[1]);
             return Pair.create(maskAndState[0], maskAndState[1]);
         }
@@ -561,7 +593,6 @@ final class WiredAccessoryManager implements WiredAccessoryCallbacks {
      * if {@code name=1}  or false if {}@code name=0} is contained in {@code state}.
      */
     private static void updateBit(int[] maskAndState, int position, String state, String name) {
-        maskAndState[0] |= position;
         if (state.contains(name + "=1")) {
             maskAndState[0] |= position;
             maskAndState[1] |= position;
