From f2eed580a03aad436bdf7e1c9da2bd2ae52cc6f7 Mon Sep 17 00:00:00 2001
From: Vaciller <wyc_19533480830@outlook.com>
Date: Fri, 10 Jul 2026 17:25:02 +0800
Subject: [PATCH] feat: customize battery meter display and remove alert state

Change-Id: I449b16a0fcda985b29a4ca57ccc6ce2524554ff9

diff --git a/packages/SystemUI/src/com/android/systemui/battery/BatteryMeterView.java b/packages/SystemUI/src/com/android/systemui/battery/BatteryMeterView.java
index 1637fbfa110a..0002ae95f476 100644
--- a/packages/SystemUI/src/com/android/systemui/battery/BatteryMeterView.java
+++ b/packages/SystemUI/src/com/android/systemui/battery/BatteryMeterView.java
@@ -82,10 +82,8 @@ public class BatteryMeterView extends LinearLayout implements DarkReceiver {
     private boolean mDisplayShieldEnabled;
     // Error state where we know nothing about the current battery state
     private boolean mBatteryStateUnknown;
-    private boolean mBatteryStateAlert;
     // Lazily-loaded since this is expected to be a rare-if-ever state
     private Drawable mUnknownStateDrawable;
-    private Drawable mAlertStateDrawable;
 
     private DualToneHandler mDualToneHandler;
 
@@ -206,14 +204,8 @@ public class BatteryMeterView extends LinearLayout implements DarkReceiver {
     public void onBatteryLevelChanged(@IntRange(from = 0, to = 100) int level, boolean pluggedIn) {
         mDrawable.setCharging(pluggedIn);
         mDrawable.setBatteryLevel(level);
-        mLevel = level;
-
-        boolean chargingChanged = mCharging != pluggedIn;
         mCharging = pluggedIn;
-        if (chargingChanged) {
-            switchBatteryIcon();
-        }
-
+        mLevel = level;
         updatePercentText();
     }
 
@@ -375,15 +367,6 @@ public class BatteryMeterView extends LinearLayout implements DarkReceiver {
         return mUnknownStateDrawable;
     }
 
-    private Drawable getAlertStateDrawable() {
-        if (mAlertStateDrawable == null) {
-            mAlertStateDrawable = mContext.getDrawable(R.drawable.ic_battery_alert);
-            mAlertStateDrawable.setTint(mTextColor);
-        }
-
-        return mAlertStateDrawable;
-    }
-
     void onBatteryUnknownStateChanged(boolean isUnknown) {
         if (mBatteryStateUnknown == isUnknown) {
             return;
@@ -392,32 +375,13 @@ public class BatteryMeterView extends LinearLayout implements DarkReceiver {
         mBatteryStateUnknown = isUnknown;
         updateContentDescription();
 
-        switchBatteryIcon();
-        updateShowPercent();
-    }
-
-    void onBatteryAlertStateChanged(boolean isAlert) {
-        if (mBatteryStateAlert == isAlert) {
-            return;
-        }
-
-        mBatteryStateAlert = isAlert;
-        updateContentDescription();
-
-        switchBatteryIcon();
-        updateShowPercent();
-    }
-
-    private void switchBatteryIcon() {
         if (mBatteryStateUnknown) {
             mBatteryIconView.setImageDrawable(getUnknownStateDrawable());
-        } else if (mCharging) {
-            mBatteryIconView.setImageDrawable(mDrawable);
-        } else if (mBatteryStateAlert) {
-            mBatteryIconView.setImageDrawable(getAlertStateDrawable());
         } else {
             mBatteryIconView.setImageDrawable(mDrawable);
         }
+
+        updateShowPercent();
     }
 
     /**
@@ -498,10 +462,6 @@ public class BatteryMeterView extends LinearLayout implements DarkReceiver {
         if (mUnknownStateDrawable != null) {
             mUnknownStateDrawable.setTint(singleToneColor);
         }
-
-        if (mAlertStateDrawable != null) {
-            mAlertStateDrawable.setTint(singleToneColor);
-        }
     }
 
     public void dump(PrintWriter pw, String[] args) {
@@ -512,7 +472,6 @@ public class BatteryMeterView extends LinearLayout implements DarkReceiver {
         pw.println("    mBatteryPercentView.getText(): " + percent);
         pw.println("    mTextColor: #" + Integer.toHexString(mTextColor));
         pw.println("    mBatteryStateUnknown: " + mBatteryStateUnknown);
-        pw.println("    mBatteryStateAlert: " + mBatteryStateAlert);
         pw.println("    mLevel: " + mLevel);
         pw.println("    mMode: " + mShowPercentMode);
     }
diff --git a/packages/SystemUI/src/com/android/systemui/battery/BatteryMeterViewController.java b/packages/SystemUI/src/com/android/systemui/battery/BatteryMeterViewController.java
index f2c056ebd57f..f4ec33ad24b5 100644
--- a/packages/SystemUI/src/com/android/systemui/battery/BatteryMeterViewController.java
+++ b/packages/SystemUI/src/com/android/systemui/battery/BatteryMeterViewController.java
@@ -94,11 +94,6 @@ public class BatteryMeterViewController extends ViewController<BatteryMeterView>
                 public void onIsOverheatedChanged(boolean isOverheated) {
                     mView.onIsOverheatedChanged(isOverheated);
                 }
-
-                @Override
-                public void onBatteryAlertStateChanged(boolean isAlert) {
-                    mView.onBatteryAlertStateChanged(isAlert);
-                }
             };
 
     private final UserTracker.Callback mUserChangedCallback =
diff --git a/packages/SystemUI/src/com/android/systemui/statusbar/policy/BatteryController.java b/packages/SystemUI/src/com/android/systemui/statusbar/policy/BatteryController.java
index fd8f303baf40..d10d7cf8460b 100644
--- a/packages/SystemUI/src/com/android/systemui/statusbar/policy/BatteryController.java
+++ b/packages/SystemUI/src/com/android/systemui/statusbar/policy/BatteryController.java
@@ -158,8 +158,6 @@ public interface BatteryController extends DemoMode,
 
         default void onIsOverheatedChanged(boolean isOverheated) {
         }
-        default void onBatteryAlertStateChanged(boolean isAlert) {
-        }
     }
 
     /**
diff --git a/packages/SystemUI/src/com/android/systemui/statusbar/policy/BatteryControllerImpl.java b/packages/SystemUI/src/com/android/systemui/statusbar/policy/BatteryControllerImpl.java
index 7e2f53eac3c2..654ba04eba7a 100644
--- a/packages/SystemUI/src/com/android/systemui/statusbar/policy/BatteryControllerImpl.java
+++ b/packages/SystemUI/src/com/android/systemui/statusbar/policy/BatteryControllerImpl.java
@@ -16,7 +16,6 @@
 
 package com.android.systemui.statusbar.policy;
 
-import static android.os.BatteryManager.BATTERY_HEALTH_DEAD;
 import static android.os.BatteryManager.BATTERY_HEALTH_OVERHEAT;
 import static android.os.BatteryManager.BATTERY_HEALTH_UNKNOWN;
 import static android.os.BatteryManager.EXTRA_HEALTH;
@@ -89,7 +88,6 @@ public class BatteryControllerImpl extends BroadcastReceiver implements BatteryC
     private int mPluggedChargingSource;
     protected boolean mCharging;
     private boolean mStateUnknown = false;
-    private boolean mStateAlert = false;
     private boolean mCharged;
     protected boolean mPowerSave;
     private boolean mAodPowerSave;
@@ -166,7 +164,6 @@ public class BatteryControllerImpl extends BroadcastReceiver implements BatteryC
         pw.print("  mIsOverheated="); pw.println(mIsOverheated);
         pw.print("  mPowerSave="); pw.println(mPowerSave);
         pw.print("  mStateUnknown="); pw.println(mStateUnknown);
-        pw.print("  mStateAlert="); pw.println(mStateAlert);
     }
 
     @Override
@@ -198,7 +195,6 @@ public class BatteryControllerImpl extends BroadcastReceiver implements BatteryC
         cb.onBatteryUnknownStateChanged(mStateUnknown);
         cb.onWirelessChargingChanged(mWirelessCharging);
         cb.onIsOverheatedChanged(mIsOverheated);
-        cb.onBatteryAlertStateChanged(mStateAlert);
     }
 
     @Override
@@ -243,11 +239,6 @@ public class BatteryControllerImpl extends BroadcastReceiver implements BatteryC
                 mIsOverheated = isOverheated;
                 fireIsOverheatedChanged();
             }
-            boolean isAlert = batteryHealth == BATTERY_HEALTH_DEAD;
-            if (isAlert != mStateAlert) {
-                mStateAlert = isAlert;
-                fireBatteryAlertStateChanged();
-            }
 
             fireBatteryLevelChanged();
         } else if (action.equals(PowerManager.ACTION_POWER_SAVE_MODE_CHANGED)) {
@@ -443,15 +434,6 @@ public class BatteryControllerImpl extends BroadcastReceiver implements BatteryC
         }
     }
 
-    private void fireBatteryAlertStateChanged() {
-        synchronized (mChangeCallbacks) {
-            final int n = mChangeCallbacks.size();
-            for (int i = 0; i < n; i++) {
-                mChangeCallbacks.get(i).onBatteryAlertStateChanged(mStateAlert);
-            }
-        }
-    }
-
     @Override
     public void dispatchDemoCommand(String command, Bundle args) {
         if (!mDemoModeController.isInDemoMode()) {
@@ -463,7 +445,6 @@ public class BatteryControllerImpl extends BroadcastReceiver implements BatteryC
         String powerSave = args.getString("powersave");
         String present = args.getString("present");
         String overheated = args.getString("overheated");
-        String alert = args.getString("alert");
         if (level != null) {
             mLevel = Math.min(Math.max(Integer.parseInt(level), 0), 100);
         }
@@ -482,10 +463,6 @@ public class BatteryControllerImpl extends BroadcastReceiver implements BatteryC
             mIsOverheated = overheated.equals("true");
             fireIsOverheatedChanged();
         }
-        if (alert != null) {
-            mStateAlert = alert.equals("true");
-            fireBatteryAlertStateChanged();
-        }
         fireBatteryLevelChanged();
     }
 
