From 738897b2d82fdaa0c68d812bed7d55d1881dd4cf Mon Sep 17 00:00:00 2001
From: Vaciller <wyc_19533480830@outlook.com>
Date: Fri, 10 Jul 2026 17:25:58 +0800
Subject: [PATCH] fix: volume dialog dismiss lifecycle and initDialog debug
 cleanup

Change-Id: Ib755b5672fa42f43c3396e467a1bdbac664ca15a

diff --git a/packages/SystemUI/src/com/android/systemui/volume/VolumeDialogImpl.java b/packages/SystemUI/src/com/android/systemui/volume/VolumeDialogImpl.java
index 7f1deb5e3009..1cfcf8cbbdcb 100644
--- a/packages/SystemUI/src/com/android/systemui/volume/VolumeDialogImpl.java
+++ b/packages/SystemUI/src/com/android/systemui/volume/VolumeDialogImpl.java
@@ -414,7 +414,6 @@ public class VolumeDialogImpl implements VolumeDialog, Dumpable,
 
     @Override
     public void destroy() {
-        Log.d(TAG, "destroy() called");
         mController.removeCallback(mControllerCallbackH);
         mHandler.removeCallbacksAndMessages(null);
         mConfigurationController.removeCallback(this);
@@ -489,7 +488,6 @@ public class VolumeDialogImpl implements VolumeDialog, Dumpable,
     }
 
     private void initDialog(int lockTaskModeState) {
-        Log.d(TAG, "initDialog: called!");
         mDialog = new CustomDialog(mContext);
 
         initDimens();
@@ -1283,7 +1281,7 @@ public class VolumeDialogImpl implements VolumeDialog, Dumpable,
     }
 
     protected void tryToRemoveCaptionsTooltip() {
-        if (mHasSeenODICaptionsTooltip && mODICaptionsTooltipView != null && mDialog != null) {
+        if (mHasSeenODICaptionsTooltip && mODICaptionsTooltipView != null) {
             ViewGroup container = mDialog.findViewById(R.id.volume_dialog_container);
             container.removeView(mODICaptionsTooltipView);
             mODICaptionsTooltipView = null;
@@ -1473,16 +1471,8 @@ public class VolumeDialogImpl implements VolumeDialog, Dumpable,
 
         mHandler.removeMessages(H.DISMISS);
         mHandler.removeMessages(H.SHOW);
-
-        boolean showingStateInconsistent = !mShowing && mDialog != null && mDialog.isShowing();
-        // If incorrectly assuming dialog is not showing, continue and make the state consistent.
-        if (showingStateInconsistent) {
-            Log.d(TAG, "dismissH: volume dialog possible in inconsistent state:"
-                    + "mShowing=" + mShowing + ", mDialog==null?" + (mDialog == null));
-        }
-        if (mIsAnimatingDismiss && !showingStateInconsistent) {
-            Log.d(TAG, "dismissH: skipping dismiss because isAnimatingDismiss is true"
-                    + " and showingStateInconsistent is false");
+        if (mIsAnimatingDismiss) {
+            Log.d(TAG, "dismissH: isAnimatingDismiss");
             Trace.endSection();
             return;
         }
@@ -1500,12 +1490,8 @@ public class VolumeDialogImpl implements VolumeDialog, Dumpable,
                 .setDuration(mDialogHideAnimationDurationMs)
                 .setInterpolator(new SystemUIInterpolators.LogAccelerateInterpolator())
                 .withEndAction(() -> mHandler.postDelayed(() -> {
-                    if (mController != null) {
-                        mController.notifyVisible(false);
-                    }
-                    if (mDialog != null) {
-                        mDialog.dismiss();
-                    }
+                    mController.notifyVisible(false);
+                    mDialog.dismiss();
                     tryToRemoveCaptionsTooltip();
                     mIsAnimatingDismiss = false;
 
