From 76c6a289dba486471937792aebda212327c69779 Mon Sep 17 00:00:00 2001
From: Vaciller <wyc_19533480830@outlook.com>
Date: Fri, 10 Jul 2026 17:20:13 +0800
Subject: [PATCH] feat: add sendAppPrivateCommand and onPrivCmdToApp for
 HDMI-IN

Change-Id: Ic41b8af722d033df27cee737177ca8b14a3f7d44

diff --git a/media/java/android/media/tv/ITvInputHardware.aidl b/media/java/android/media/tv/ITvInputHardware.aidl
index 94c1013a837e..8e393b513e8a 100644
--- a/media/java/android/media/tv/ITvInputHardware.aidl
+++ b/media/java/android/media/tv/ITvInputHardware.aidl
@@ -17,6 +17,7 @@
 package android.media.tv;
 
 import android.media.tv.TvStreamConfig;
+import android.os.Bundle;
 import android.view.KeyEvent;
 import android.view.Surface;
 
@@ -55,4 +56,6 @@ interface ITvInputHardware {
      */
     void overrideAudioSink(int audioType, String audioAddress, int samplingRate, int channelMask,
             int format);
+
+    void sendAppPrivateCommand(String action, in Bundle data);
 }
diff --git a/media/java/android/media/tv/ITvInputHardwareCallback.aidl b/media/java/android/media/tv/ITvInputHardwareCallback.aidl
index 870883b2a28b..5a3d277e276a 100644
--- a/media/java/android/media/tv/ITvInputHardwareCallback.aidl
+++ b/media/java/android/media/tv/ITvInputHardwareCallback.aidl
@@ -17,6 +17,7 @@
 package android.media.tv;
 
 import android.media.tv.TvStreamConfig;
+import android.os.Bundle;
 
 /**
  * @hide
@@ -24,4 +25,5 @@ import android.media.tv.TvStreamConfig;
 oneway interface ITvInputHardwareCallback {
     void onReleased();
     void onStreamConfigChanged(in TvStreamConfig[] configs);
+    void onPrivCmdToApp(String action, in Bundle bundle);
 }
diff --git a/media/java/android/media/tv/TvInputManager.java b/media/java/android/media/tv/TvInputManager.java
index 149c2f471a4c..2ad7e13a3c9b 100644
--- a/media/java/android/media/tv/TvInputManager.java
+++ b/media/java/android/media/tv/TvInputManager.java
@@ -1129,6 +1129,8 @@ public final class TvInputManager {
          * @param configs The new {@link TvStreamConfig}s.
          */
         public abstract void onStreamConfigChanged(TvStreamConfig[] configs);
+
+        public abstract void onPrivCmdToApp(@NonNull String action, @NonNull Bundle bundle);
     }
 
     /**
@@ -2185,6 +2187,16 @@ public final class TvInputManager {
                                 Binder.restoreCallingIdentity(identity);
                             }
                 }
+
+                @Override
+                public void onPrivCmdToApp(String action, Bundle bundle) {
+                            final long identity = Binder.clearCallingIdentity();
+                            try {
+                                executor.execute(() -> callback.onPrivCmdToApp(action, bundle));
+                            } finally {
+                                Binder.restoreCallingIdentity(identity);
+                            }
+                }
                     }, info, mUserId, tvInputSessionId, priorityHint);
             if (hardware == null) {
                 return null;
@@ -3307,6 +3319,14 @@ public final class TvInputManager {
             }
         }
 
+        public void sendAppPrivateCommand(@NonNull String action, @NonNull Bundle data) {
+            try {
+                mInterface.sendAppPrivateCommand(action, data);
+            } catch (RemoteException e) {
+                throw new RuntimeException(e);
+            }
+        }
+
         /** @removed */
         @SystemApi
         public boolean dispatchKeyEventToHdmi(KeyEvent event) {
diff --git a/media/java/android/media/tv/TvView.java b/media/java/android/media/tv/TvView.java
index ff3d06c5bd69..dea84ad78373 100644
--- a/media/java/android/media/tv/TvView.java
+++ b/media/java/android/media/tv/TvView.java
@@ -72,7 +72,7 @@ import java.util.Queue;
  */
 public class TvView extends ViewGroup {
     private static final String TAG = "TvView";
-    private static final boolean DEBUG = false;
+    private static final boolean DEBUG = true;
 
     private static final int ZORDER_MEDIA = 0;
     private static final int ZORDER_MEDIA_OVERLAY = 1;
@@ -128,12 +128,18 @@ public class TvView extends ViewGroup {
 
         @Override
         public void surfaceCreated(SurfaceHolder holder) {
+            if (DEBUG) {
+                Log.d(TAG, "surfaceCreated(holder=" + holder + ")");
+            }
             mSurface = holder.getSurface();
             setSessionSurface(mSurface);
         }
 
         @Override
         public void surfaceDestroyed(SurfaceHolder holder) {
+            if (DEBUG) {
+                Log.d(TAG, "surfaceDestroyed(holder=" + holder + ")");
+            }
             mSurface = null;
             mSurfaceChanged = false;
             setSessionSurface(null);
@@ -334,6 +340,7 @@ public class TvView extends ViewGroup {
             }
         }
         if (mSessionCallback != null && TextUtils.equals(mSessionCallback.mInputId, inputId)) {
+            if (DEBUG) Log.d(TAG, "tune(" + channelUri + ") mSession=" + mSession);
             if (mSession != null) {
                 mSession.tune(channelUri, params);
             } else {
@@ -355,9 +362,11 @@ public class TvView extends ViewGroup {
             // is obsolete and should ignore it.
             mSessionCallback = new MySessionCallback(inputId, channelUri, params);
             if (mTvInputManager != null) {
+                if (DEBUG) Log.d(TAG, "tune(" + channelUri + ") createSession inputId=" + inputId);
                 mTvInputManager.createSession(inputId, mSessionCallback, mHandler);
             }
         }
+        if (DEBUG) Log.w(TAG, "tune(" + channelUri + ") finish");
     }
 
     /**
@@ -366,13 +375,14 @@ public class TvView extends ViewGroup {
      * <p>This method is primarily used to un-tune the current TvView.
      */
     public void reset() {
-        if (DEBUG) Log.d(TAG, "reset()");
+        if (DEBUG) Log.d(TAG, "reset() start");
         synchronized (sMainTvViewLock) {
             if (this == sMainTvView.get()) {
                 sMainTvView = NULL_TV_VIEW;
             }
         }
         resetInternal();
+        if (DEBUG) Log.d(TAG, "reset() finish");
     }
 
     private void resetInternal() {
@@ -383,6 +393,7 @@ public class TvView extends ViewGroup {
             removeSessionOverlayView();
             mUseRequestedSurfaceLayout = false;
             mSession.release();
+            if (DEBUG) Log.d(TAG, "resetInternal finish mSession.release()");
             mSession = null;
             resetSurfaceView();
         }
@@ -730,12 +741,18 @@ public class TvView extends ViewGroup {
     protected void onAttachedToWindow() {
         super.onAttachedToWindow();
         createSessionOverlayView();
+        if (DEBUG) {
+            Log.d(TAG, "onAttachedToWindow");
+        }
     }
 
     @Override
     protected void onDetachedFromWindow() {
         removeSessionOverlayView();
         super.onDetachedFromWindow();
+        if (DEBUG) {
+            Log.d(TAG, "onDetachedFromWindow");
+        }
     }
 
     @Override
@@ -811,6 +828,7 @@ public class TvView extends ViewGroup {
     }
 
     private void resetSurfaceView() {
+        if (DEBUG) Log.d(TAG, "resetSurfaceView() start");
         if (mSurfaceView != null) {
             mSurfaceView.getHolder().removeCallback(mSurfaceHolderCallback);
             removeView(mSurfaceView);
@@ -831,12 +849,14 @@ public class TvView extends ViewGroup {
             mSurfaceView.setZOrderOnTop(true);
         }
         addView(mSurfaceView);
+        if (DEBUG) Log.d(TAG, "resetSurfaceView() finish");
     }
 
     private void setSessionSurface(Surface surface) {
         if (mSession == null) {
             return;
         }
+        if (DEBUG) Log.d(TAG, "setSessionSurface(surface=" + surface + ")");
         mSession.setSurface(surface);
     }
 
