From 6279f55c1108cbaa3a371f1cebb82347028940e0 Mon Sep 17 00:00:00 2001
From: Vaciller <wyc_19533480830@outlook.com>
Date: Fri, 10 Jul 2026 17:20:55 +0800
Subject: [PATCH] feat: add HDMI-IN Intent action, EDID writer, and config
 entries

Change-Id: Ifbcf8edd1ba19cc7b65f0e10c1fd8f6d43996e9e

diff --git a/core/java/android/content/Intent.java b/core/java/android/content/Intent.java
index 809dc3c41188..224c31ebe751 100644
--- a/core/java/android/content/Intent.java
+++ b/core/java/android/content/Intent.java
@@ -5067,6 +5067,13 @@ public class Intent implements Parcelable, Cloneable {
     public static final String ACTION_PACKAGE_NEEDS_INTEGRITY_VERIFICATION =
             "android.intent.action.PACKAGE_NEEDS_INTEGRITY_VERIFICATION";
 
+    /**
+     * Broadcast Action: for hdmiin
+     * @hide
+     */
+    public static final String ACTION_HDMIIN_RK_PRIV_CMD =
+            "com.rk.hdmiin.privcmdtoapp";
+
     /**
      * Broadcast Action: Start the foreground service manager.
      *
diff --git a/core/res/res/values/config.xml b/core/res/res/values/config.xml
index 514083cefca9..9891409a08bb 100644
--- a/core/res/res/values/config.xml
+++ b/core/res/res/values/config.xml
@@ -572,6 +572,9 @@
          docked if the dock is configured to enable the accelerometer. -->
     <bool name="config_supportAutoRotation">true</bool>
 
+    <!-- If true, allows rotation resolver service to help resolve screen rotation. -->
+    <bool name="config_allowRotationResolver">true</bool>
+
     <!-- If true, the screen can be rotated via the accelerometer in all 4
          rotations as the default behavior. -->
     <bool name="config_allowAllRotations">false</bool>
diff --git a/core/res/res/values/symbols.xml b/core/res/res/values/symbols.xml
index 17a8a76bd590..01c6b5509cb7 100644
--- a/core/res/res/values/symbols.xml
+++ b/core/res/res/values/symbols.xml
@@ -1754,6 +1754,7 @@
   <java-symbol type="bool" name="config_perDisplayFocusEnabled" />
   <java-symbol type="bool" name="config_showNavigationBar" />
   <java-symbol type="bool" name="config_supportAutoRotation" />
+  <java-symbol type="bool" name="config_allowRotationResolver" />
   <java-symbol type="bool" name="config_dockedStackDividerFreeSnapMode" />
   <java-symbol type="dimen" name="docked_stack_divider_thickness" />
   <java-symbol type="dimen" name="docked_stack_divider_insets" />
diff --git a/media/java/android/media/audio/common/AidlConversion.java b/media/java/android/media/audio/common/AidlConversion.java
index 491a8ec2c130..bf4d30be101f 100644
--- a/media/java/android/media/audio/common/AidlConversion.java
+++ b/media/java/android/media/audio/common/AidlConversion.java
@@ -741,10 +741,18 @@ public class AidlConversion {
                 aidl.type = AudioDeviceType.IN_DEVICE;
                 aidl.connection = AudioDeviceDescription.CONNECTION_HDMI;
                 break;
+            case AudioSystem.DEVICE_IN_HDMI_1:
+                aidl.type = AudioDeviceType.IN_DEVICE;
+                aidl.connection = "VX_ROCKCHIP_HDMI0";
+                break;
             case AudioSystem.DEVICE_OUT_HDMI:
                 aidl.type = AudioDeviceType.OUT_DEVICE;
                 aidl.connection = AudioDeviceDescription.CONNECTION_HDMI;
                 break;
+            case AudioSystem.DEVICE_OUT_HDMI_1:
+                aidl.type = AudioDeviceType.OUT_DEVICE;
+                aidl.connection = "VX_ROCKCHIP_HDMI0";
+                break;
             case AudioSystem.DEVICE_IN_REMOTE_SUBMIX:
                 aidl.type = AudioDeviceType.IN_SUBMIX;
                 break;
@@ -805,6 +813,10 @@ public class AidlConversion {
                 aidl.type = AudioDeviceType.OUT_DEVICE;
                 aidl.connection = AudioDeviceDescription.CONNECTION_SPDIF;
                 break;
+            case AudioSystem.DEVICE_OUT_SPDIF_1:
+                aidl.type = AudioDeviceType.OUT_DEVICE;
+                aidl.connection = "VX_ROCKCHIP_SPDIF0";
+                break;
             case AudioSystem.DEVICE_IN_BLUETOOTH_A2DP:
                 aidl.type = AudioDeviceType.IN_DEVICE;
                 aidl.connection = AudioDeviceDescription.CONNECTION_BT_A2DP;
diff --git a/services/java/com/android/server/SystemServer.java b/services/java/com/android/server/SystemServer.java
index 46141940a907..13071581a1a3 100644
--- a/services/java/com/android/server/SystemServer.java
+++ b/services/java/com/android/server/SystemServer.java
@@ -111,6 +111,7 @@ import com.android.server.art.ArtManagerLocal;
 import com.android.server.attention.AttentionManagerService;
 import com.android.server.audio.AudioService;
 import com.android.server.biometrics.AuthService;
+import com.android.server.audio.RkAudioSettingService;
 import com.android.server.biometrics.BiometricService;
 import com.android.server.biometrics.sensors.face.FaceService;
 import com.android.server.biometrics.sensors.fingerprint.FingerprintService;
@@ -218,6 +219,7 @@ import dalvik.system.VMRuntime;
 
 import java.io.File;
 import java.io.FileDescriptor;
+import java.io.FileOutputStream;
 import java.io.IOException;
 import java.io.PrintWriter;
 import java.text.SimpleDateFormat;
@@ -1531,6 +1533,16 @@ public final class SystemServer implements Dumpable {
             mSystemServiceManager.startService(DropBoxManagerService.class);
             t.traceEnd();
 
+            //-----rk-code-------
+            t.traceBegin("StartRKBoxService");
+            try {
+                ServiceManager.addService("rkbox_management", new RKBoxManagementService(context));
+            } catch (Throwable e) {
+                Slog.e(TAG, "Failure starting RKBoxManagement Service", e);
+            }
+            t.traceEnd();
+            //------------------
+
             // Grants default permissions and defines roles
             t.traceBegin("StartRoleManagerService");
             LocalManagerRegistry.addManager(RoleServicePlatformHelper.class,
@@ -2042,6 +2054,25 @@ public final class SystemServer implements Dumpable {
             }
             t.traceEnd();
 
+            // $_rbox_$_modify_$_aisx: added 2017-06-27, add RkDisplayDeviceManagementService
+            if (mPackageManager.hasSystemFeature(PackageManager.FEATURE_ROCKCHIP_DISPLAY)) {
+                Slog.i(TAG, "addService drm_device_management");
+                try {
+                    ServiceManager.addService("drm_device_management",
+                            new RkDisplayDeviceManagementService(context));
+                } catch (Throwable e) {
+                    Slog.e(TAG, "Failure starting kDisplayDeviceManagement Service", e);
+                }
+            }
+            if (mPackageManager.hasSystemFeature(PackageManager.FEATURE_ROCKCHIP_AUDIO)) {
+                Slog.i(TAG, "addService rockchip_audio_setting");
+                try {
+                    ServiceManager.addService("rockchip_audio_setting", new RkAudioSettingService(context));
+                } catch (Throwable e) {
+                    Slog.e(TAG, "Failure starting RkAudioSettingManager Service", e);
+                }
+            }
+
             t.traceBegin("StartNotificationManager");
             mSystemServiceManager.startService(NotificationManagerService.class);
             SystemNotificationChannels.removeDeprecated(context);
@@ -2405,6 +2436,7 @@ public final class SystemServer implements Dumpable {
                     || mPackageManager.hasSystemFeature(PackageManager.FEATURE_LEANBACK)) {
                 t.traceBegin("StartTvInputManager");
                 mSystemServiceManager.startService(TvInputManagerService.class);
+                writeHdmiRxEdid();
                 t.traceEnd();
             }
 
@@ -3082,6 +3114,26 @@ public final class SystemServer implements Dumpable {
         t.traceEnd(); // startApexServices
     }
 
+    private void writeHdmiRxEdid() {
+        FileOutputStream file = null;
+        try{
+            String hdmiinVersion = SystemProperties.get("persist.sys.hdmirx.edid", "1");
+            Slog.v(TAG, "write hdmirx edid value " + hdmiinVersion);
+            file = new FileOutputStream("sys/class/hdmirx/hdmirx/edid");
+            file.write(hdmiinVersion.getBytes());
+            file.flush();
+        } catch(Exception e){
+            e.printStackTrace();
+        } finally {
+            if (null != file) {
+                try{
+                    file.close();
+                } catch(Exception e1) {
+                }
+            }
+        }
+    }
+
     private boolean deviceHasConfigString(@NonNull Context context, @StringRes int resId) {
         String serviceName = context.getString(resId);
         return !TextUtils.isEmpty(serviceName);
