From afb3dcf38ef8af9edd9a4ea25efe3b19e631f99e Mon Sep 17 00:00:00 2001
From: Vaciller <wyc_19533480830@outlook.com>
Date: Fri, 10 Jul 2026 17:21:08 +0800
Subject: [PATCH] feat: add bundled_persist and bundled_uninstall app
 directories

Change-Id: If4261b395d96c02a84f98078f83009d5c36f5ecb

diff --git a/core/java/android/os/Environment.java b/core/java/android/os/Environment.java
index 5177cb4f8549..cbf2556648ab 100644
--- a/core/java/android/os/Environment.java
+++ b/core/java/android/os/Environment.java
@@ -65,6 +65,8 @@ public class Environment {
     private static final String ENV_PRODUCT_ROOT = "PRODUCT_ROOT";
     private static final String ENV_SYSTEM_EXT_ROOT = "SYSTEM_EXT_ROOT";
     private static final String ENV_APEX_ROOT = "APEX_ROOT";
+    private static final String ENV_PREBUNDLED_UNINSTALL_BACK_ROOT = "PREBUNDLED_UNINSTALL_BACK_ROOT";
+    private static final String ENV_PREBUNDLED_UNINSTALL_GONE_ROOT = "PREBUNDLED_UNINSTALL_GONE_ROOT";
 
     /** {@hide} */
     public static final String DIR_ANDROID = "Android";
@@ -109,6 +111,10 @@ public class Environment {
             "/system_ext");
     private static final File DIR_APEX_ROOT = getDirectory(ENV_APEX_ROOT,
             "/apex");
+    private static final File DIR_PREBUNDLED_UNINSTALL_BACK_ROOT = getDirectory(
+            ENV_PREBUNDLED_UNINSTALL_BACK_ROOT, "/odm/bundled_uninstall_back-app");
+    private static final File DIR_PREBUNDLED_UNINSTALL_GONE_ROOT = getDirectory(
+            ENV_PREBUNDLED_UNINSTALL_GONE_ROOT, "/odm/bundled_uninstall_gone-app");
 
     /**
      * Scoped Storage is on by default. However, it is not strictly enforced and there are multiple
@@ -1524,6 +1530,22 @@ public class Environment {
         return cur;
     }
 
+    /**
+     * Return the root directory for "prebundled" apps.  These apps will be installed directly
+     * from this partition but will not be marked as system apps and will hence be uninstallable.
+     * @hide
+     */
+    public static File getPrebundledUninstallBackDirectory() {
+        return DIR_PREBUNDLED_UNINSTALL_BACK_ROOT;
+    }
+
+    /**
+     * @hide
+     */
+    public static File getPrebundledUninstallGoneDirectory() {
+        return DIR_PREBUNDLED_UNINSTALL_GONE_ROOT;
+    }
+
     /**
      * If the given path exists on emulated external storage, return the
      * translated backing path hosted on internal storage. This bypasses any
diff --git a/data/etc/privapp-permissions-platform.xml b/data/etc/privapp-permissions-platform.xml
index e1c9b3c121f8..4eb270a7fd2a 100644
--- a/data/etc/privapp-permissions-platform.xml
+++ b/data/etc/privapp-permissions-platform.xml
@@ -39,6 +39,7 @@ applications that come with the platform
     </privapp-permissions>
 
     <privapp-permissions package="com.android.externalstorage">
+        <permission name="android.permission.MANAGE_USERS"/>
         <permission name="android.permission.MOUNT_UNMOUNT_FILESYSTEMS"/>
         <permission name="android.permission.WRITE_MEDIA_STORAGE"/>
         <permission name="android.permission.MANAGE_EXTERNAL_STORAGE"/>
diff --git a/packages/ExternalStorageProvider/AndroidManifest.xml b/packages/ExternalStorageProvider/AndroidManifest.xml
index 9b43d9f22390..e9b77360ecdc 100644
--- a/packages/ExternalStorageProvider/AndroidManifest.xml
+++ b/packages/ExternalStorageProvider/AndroidManifest.xml
@@ -6,6 +6,7 @@
     <uses-permission android:name="android.permission.WRITE_MEDIA_STORAGE" />
     <uses-permission android:name="android.permission.MANAGE_EXTERNAL_STORAGE" />
     <uses-permission android:name="android.permission.MOUNT_UNMOUNT_FILESYSTEMS" />
+    <uses-permission android:name="android.permission.MANAGE_USERS" />
 
     <application android:label="@string/app_label">
         <provider
diff --git a/services/core/java/com/android/server/pm/PackageManagerService.java b/services/core/java/com/android/server/pm/PackageManagerService.java
index 47860373156b..1877fb56f8d8 100644
--- a/services/core/java/com/android/server/pm/PackageManagerService.java
+++ b/services/core/java/com/android/server/pm/PackageManagerService.java
@@ -115,6 +115,7 @@ import android.content.pm.VerifierDeviceIdentity;
 import android.content.pm.VersionedPackage;
 import android.content.pm.overlay.OverlayPaths;
 import android.content.pm.parsing.PackageLite;
+import android.content.res.Configuration;
 import android.content.res.Resources;
 import android.database.ContentObserver;
 import android.graphics.Bitmap;
@@ -248,10 +249,13 @@ import java.io.ByteArrayInputStream;
 import java.io.ByteArrayOutputStream;
 import java.io.File;
 import java.io.FileDescriptor;
+import java.io.FileInputStream;
 import java.io.FileNotFoundException;
+import java.io.FileOutputStream;
 import java.io.IOException;
 import java.io.InputStream;
 import java.io.PrintWriter;
+import java.io.RandomAccessFile;
 import java.lang.annotation.Retention;
 import java.lang.annotation.RetentionPolicy;
 import java.security.MessageDigest;
@@ -276,6 +280,9 @@ import java.util.concurrent.atomic.AtomicInteger;
 import java.util.concurrent.atomic.AtomicReference;
 import java.util.function.Consumer;
 
+import org.xmlpull.v1.XmlPullParser;
+import org.xmlpull.v1.XmlPullParserException;
+
 /**
  * Keep track of all those APKs everywhere.
  * <p>
@@ -344,6 +351,9 @@ public class PackageManagerService implements PackageSender, TestUtilityService
 
     static final boolean HIDE_EPHEMERAL_APIS = false;
 
+    //for UiMode Debug
+    private static final boolean DEBUG_UIMODE = Log.isLoggable(TAG, Log.DEBUG);
+
     static final String PRECOMPILE_LAYOUTS = "pm.precompile_layouts";
 
     private static final int RADIO_UID = Process.PHONE_UID;
@@ -378,6 +388,8 @@ public class PackageManagerService implements PackageSender, TestUtilityService
     static final int SCAN_AS_ODM = 1 << 22;
     static final int SCAN_AS_APK_IN_APEX = 1 << 23;
     static final int SCAN_DROP_CACHE = 1 << 24;
+    static final int SCAN_AS_PREINSTALL = 1<<25;
+    static final int SCAN_AS_PREBUNDLED_DIR = 1<<26;
 
     @IntDef(flag = true, prefix = { "SCAN_" }, value = {
             SCAN_NO_DEX,
@@ -523,6 +535,12 @@ public class PackageManagerService implements PackageSender, TestUtilityService
 
     private static final String COMPANION_PACKAGE_NAME = "com.android.companiondevicemanager";
 
+    public static final String BUNDLED_PERSIST_DIR = "/odm/bundled_persist-app";
+
+    public static final String BUNDLED_UNINSTALL_GONE_DIR = "/odm/bundled_uninstall_gone-app";
+
+    public static final String DELETE_APK_FILE = "/cache/recovery/last_deleteApkFile.dat";
+
     // Compilation reasons.
     public static final int REASON_FIRST_BOOT = 0;
     public static final int REASON_BOOT_AFTER_OTA = 1;
@@ -4019,6 +4037,10 @@ public class PackageManagerService implements PackageSender, TestUtilityService
         mPrepareAppDataFuture = null;
     }
 
+    private Map<String, Integer> mPackageUiModeConfigMap = new HashMap<String, Integer>();
+    private static String VENDOR_CONFIG = "/vendor/etc/package_uimode_config.xml";
+    private static String DATA_CONFIG = "/data/system/package_uimode_config.xml";
+
     public void systemReady() {
         PackageManagerServiceUtils.enforceSystemOrRoot(
                 "Only the system can claim the system is ready");
@@ -5408,6 +5430,33 @@ public class PackageManagerService implements PackageSender, TestUtilityService
             updateComponentLabelIcon(componentName, null, null, userId);
         }
 
+        /**
+         * @hide
+         */
+        public int getPackagePerformanceMode(String pkgName) {
+            for (int i = 0; i < mSettings.mPerformancePackages.size(); i++) {
+                if (pkgName.toLowerCase().contains(mSettings.mPerformancePackages.get(i).name.toLowerCase())) {
+                    return mSettings.mPerformancePackages.get(i).mode;
+                }
+            }
+            return 0;
+        }
+
+        /**
+         * @hide
+         */
+        public void setPackagePerformanceMode(String pkgName) {
+            for (int i = 0; i < mSettings.mPerformancePackages.size(); i++) {
+                if (pkgName.toLowerCase().contains(mSettings.mPerformancePackages.get(i).name.toLowerCase())) {
+                    Slog.w(TAG, "Already in performance mode: " + pkgName);
+                    return;
+                }
+            }
+            PackagePerformanceSetting app = new PackagePerformanceSetting(pkgName, 1 /* enable performance mode */);
+            mSettings.mPerformancePackages.add(app);
+            Slog.i(TAG, "Added in performance mode: " + pkgName);
+        }
+
         @Override
         public void sendDeviceCustomizationReadyBroadcast() {
             mContext.enforceCallingPermission(Manifest.permission.SEND_DEVICE_CUSTOMIZATION_READY,
@@ -6032,6 +6081,230 @@ public class PackageManagerService implements PackageSender, TestUtilityService
                     getPerUidReadTimeouts(snapshot)
             ).doDump(snapshot, fd, pw, args);
         }
+
+        /**
+         * @hide
+         */
+        public void setPackageUiModeType(String packageName, int oldUiMode, int newUiMode) {
+            RandomAccessFile randomAccessFile = null;
+            try {
+                File configFilter = new File(DATA_CONFIG);
+                randomAccessFile = new RandomAccessFile(configFilter, "rw");
+                String line = null;
+                long lastPoint = 0;
+                StringBuilder totalStr = new StringBuilder();
+                while ((line = randomAccessFile.readLine()) != null) {
+                    long point = randomAccessFile.getFilePointer();
+                    if (line.contains(packageName)
+                            && line.contains(String.format("uiMode=\"%d\"", oldUiMode))) {
+                        String str = line.replace(String.format("uiMode=\"%d\"", oldUiMode),
+                                String.format("uiMode=\"%d\"", newUiMode));
+                        randomAccessFile.seek(lastPoint);
+                        randomAccessFile.writeBytes(str);
+                        totalStr.append(str);
+                        lastPoint = point;
+                        continue;
+                    }
+                    lastPoint = point;
+                    totalStr.append(line);
+                }
+                if (!totalStr.toString().contains(packageName)) {
+                    randomAccessFile.seek(lastPoint - 26);
+                    randomAccessFile.writeBytes(String.format(
+                            "    <app package=\"%s\" uiMode=\"%d\"/>\n</ui-mode-package-config>\n",
+                            packageName, newUiMode));
+                }
+            } catch (Exception e) {
+                e.printStackTrace();
+            } finally {
+                try {
+                    if (randomAccessFile != null) {
+                        randomAccessFile.close();
+                    }
+                } catch (IOException e) {
+                    e.printStackTrace();
+                }
+            }
+            File configFilter = new File(DATA_CONFIG);
+            if (configFilter.exists()) {
+                mPackageUiModeConfigMap.clear();
+                FileInputStream stream = null;
+                try {
+                    stream = new FileInputStream(configFilter);
+                    XmlPullParser xmlPullParser = Xml.newPullParser();
+                    xmlPullParser.setInput(stream, null);
+                    int type;
+                    do {
+                        type = xmlPullParser.next();
+                        if (type == XmlPullParser.START_TAG) {
+                            String tag = xmlPullParser.getName();
+                            if (DEBUG_UIMODE) {
+                                Slog.d(TAG, "getConfigMap: tag = " + tag);
+                            }
+                            if ("app".equals(tag)) {
+                                String pkgName = xmlPullParser.getAttributeValue(null, "package");
+                                String uiMode = xmlPullParser.getAttributeValue(null, "uiMode");
+                                if (DEBUG_UIMODE) {
+                                    Slog.d(TAG, "getConfigMap: pkgName = " + pkgName + ", uiMode = "
+                                            + uiMode);
+                                }
+                                if (!TextUtils.isEmpty(pkgName) && !TextUtils.isEmpty(uiMode)) {
+                                    int parseUiMode = Integer.parseInt(uiMode);
+                                    mPackageUiModeConfigMap.put(pkgName,
+                                            parseUiMode >= 0 ? parseUiMode : -1);
+                                }
+                            } else {
+                                if (DEBUG_UIMODE) {
+                                    Slog.d(TAG, "getConfigMap: , tag != app");
+                                }
+                            }
+                        }
+                    } while (type != XmlPullParser.END_DOCUMENT);
+                } catch (NullPointerException e) {
+                    Slog.w(TAG, "failed parsing " + configFilter, e);
+                } catch (NumberFormatException e) {
+                    Slog.w(TAG, "failed parsing " + configFilter, e);
+                } catch (XmlPullParserException e) {
+                    Slog.w(TAG, "failed parsing " + configFilter, e);
+                } catch (IndexOutOfBoundsException e) {
+                    Slog.w(TAG, "failed parsing " + configFilter, e);
+                } catch (IOException e) {
+                    Slog.w(TAG, "failed parsing " + configFilter, e);
+                } finally {
+                    try {
+                        if (stream != null) {
+                            stream.close();
+                        }
+                    } catch (IOException e) {
+                        Slog.w(TAG, "stream.close failed");
+                    }
+                }
+            } else {
+                if (DEBUG_UIMODE) {
+                    Slog.w(TAG, "package_uimode_config.xml is not exists");
+                }
+            }
+        }
+
+        /**
+         * @hide
+         */
+        public int getPackageUiModeType(String packageName) {
+            File configFilter = new File(DATA_CONFIG);
+            if (!configFilter.exists()) {
+                try {
+                    int byteSum = 0;
+                    int byteRead = 0;
+                    File vendorFilter = new File(VENDOR_CONFIG);
+                    configFilter.createNewFile();
+                    if (vendorFilter.exists()) {
+                        InputStream inStream = new FileInputStream(vendorFilter);
+                        FileOutputStream fs = new FileOutputStream(configFilter);
+                        byte[] buffer = new byte[1024];
+                        int length;
+                        while ((byteRead = inStream.read(buffer)) != -1) {
+                            byteSum += byteRead;
+                            fs.write(buffer, 0, byteRead);
+                        }
+                        inStream.close();
+                        fs.close();
+                    }
+                    int result = android.os.FileUtils.setPermissions(
+                            configFilter, android.os.FileUtils.S_IRWXU
+                                    | android.os.FileUtils.S_IRWXG | android.os.FileUtils.S_IRWXO,
+                            -1, -1);
+                    if (DEBUG_UIMODE) {
+                        Slog.w(TAG, "chmod file result = " + result);
+                    }
+                } catch (IOException e) {
+                    e.printStackTrace();
+                }
+            }
+            configFilter = new File(DATA_CONFIG);
+            if (mPackageUiModeConfigMap != null && mPackageUiModeConfigMap.size() <= 0
+                    && configFilter.exists()) {
+                FileInputStream stream = null;
+                try {
+                    stream = new FileInputStream(configFilter);
+                    XmlPullParser xmlPullParser = Xml.newPullParser();
+                    xmlPullParser.setInput(stream, null);
+                    int type;
+                    do {
+                        type = xmlPullParser.next();
+                        if (type == XmlPullParser.START_TAG) {
+                            String tag = xmlPullParser.getName();
+                            if (DEBUG_UIMODE) {
+                                Slog.d(TAG, "getConfigMap: tag = " + tag);
+                            }
+                            if ("app".equals(tag)) {
+                                String pkgName = xmlPullParser.getAttributeValue(null, "package");
+                                String uiMode = xmlPullParser.getAttributeValue(null, "uiMode");
+                                if (DEBUG_UIMODE) {
+                                    Slog.d(TAG, "getConfigMap: pkgName = " + pkgName + ", uiMode = "
+                                            + uiMode);
+                                }
+                                if (!TextUtils.isEmpty(pkgName) && !TextUtils.isEmpty(uiMode)) {
+                                    int parseUiMode = Integer.parseInt(uiMode);
+                                    mPackageUiModeConfigMap.put(pkgName,
+                                            parseUiMode >= 0 ? parseUiMode : -1);
+                                }
+                            } else {
+                                if (DEBUG_UIMODE) {
+                                    Slog.d(TAG, "getConfigMap: , tag != app");
+                                }
+                            }
+                        }
+                    } while (type != XmlPullParser.END_DOCUMENT);
+                } catch (NullPointerException e) {
+                    Slog.w(TAG, "failed parsing " + configFilter, e);
+                } catch (NumberFormatException e) {
+                    Slog.w(TAG, "failed parsing " + configFilter, e);
+                } catch (XmlPullParserException e) {
+                    Slog.w(TAG, "failed parsing " + configFilter, e);
+                } catch (IndexOutOfBoundsException e) {
+                    Slog.w(TAG, "failed parsing " + configFilter, e);
+                } catch (IOException e) {
+                    Slog.w(TAG, "failed parsing " + configFilter, e);
+                } finally {
+                    try {
+                        if (stream != null) {
+                            stream.close();
+                        }
+                    } catch (IOException e) {
+                        Slog.w(TAG, "stream.close failed");
+                    }
+                }
+            } else {
+                if (DEBUG_UIMODE) {
+                    Slog.w(TAG,
+                            "package_uimode_config.xml is not exists or mPackageUiModeConfigMap > 0");
+                }
+            }
+
+            if (mPackageUiModeConfigMap != null && mPackageUiModeConfigMap.size() > 0) {
+                if (!TextUtils.isEmpty(packageName)) {
+                    for (String pkgName : mPackageUiModeConfigMap.keySet()) {
+                        if (!TextUtils.isEmpty(pkgName) && packageName.equals(pkgName)) {
+                            Integer uiMode = mPackageUiModeConfigMap.get(packageName);
+                            if (uiMode != null && uiMode >= 0) {
+                                if (DEBUG_UIMODE) {
+                                    Slog.d(TAG, "fix uiMode for app package = " + packageName
+                                            + " , uiMode = " + uiMode.toString());
+                                }
+                                return uiMode;
+                            } else {
+                                continue;
+                            }
+                        } else {
+                            continue;
+                        }
+                    }
+                }
+            } else {
+                return -1;
+            }
+            return -1;
+        }
     }
 
     private class PackageManagerLocalImpl implements PackageManagerLocal {
